<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>7.3. Data &mdash; Topics in Bioinformatics 2022.07.28 documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="7.4. The neutral theory of molecular evolution" href="neutral_theory.html" />
    <link rel="prev" title="7.1. What is molecular evolution?" href="intro.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">7. Molecular Evolution</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="intro.html">7.1. What is molecular evolution?</a></li>
<li class="toctree-l2"><a class="reference internal" href="intro.html#what-is-phylogenetics">7.2. What is phylogenetics?</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">7.3. Data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#classes-of-related-sequences">7.3.1. Classes of related sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="#exercises">7.3.2. Exercises</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="neutral_theory.html">7.4. The neutral theory of molecular evolution</a></li>
<li class="toctree-l2"><a class="reference internal" href="mutation.html">7.5. Mutation – the origin of genetic variation</a></li>
<li class="toctree-l2"><a class="reference internal" href="max_likelihood.html">7.6. Statistical methods for molecular evolutionary analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="substitution_models.html">7.7. Substitution models</a></li>
<li class="toctree-l2"><a class="reference internal" href="substitution_models.html#exercises">7.8. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="hypothesis_testing.html">7.9. Hypothesis Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="hypothesis_testing.html#exercises">7.10. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="glossary.html">7.11. Glossary of terms</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">9. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">10. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">11. Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html"><span class="section-number">7. </span>Molecular Evolution</a> &raquo;</li>
      <li><span class="section-number">7.3. </span>Data</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/molevol/data.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<div class="section" id="data">
<h1><span class="section-number">7.3. </span>Data<a class="headerlink" href="#data" title="Permalink to this headline"></a></h1>
<div class="section" id="classes-of-related-sequences">
<h2><span class="section-number">7.3.1. </span>Classes of related sequences<a class="headerlink" href="#classes-of-related-sequences" title="Permalink to this headline"></a></h2>
<dl class="simple">
<dt>Homolog</dt><dd><p>Sequences descended from a common ancestor.</p>
</dd>
<dt>Ortholog</dt><dd><p>Sequences that have evolved from a common ancestor by speciation. Typically, orthologous genes in different species retain the same function. For example, the TP53 gene in human and mouse.</p>
</dd>
<dt>Paralog</dt><dd><p>Paralogs are sequences that arose from duplication within the same genome. For example, multi-gene families such as the haemoglobins.</p>
</dd>
</dl>
<div class="figure align-default" id="id35">
<a class="reference internal image-reference" href="../_images/homologs.png"><img alt="../_images/homologs.png" src="../_images/homologs.png" style="width: 623.0px; height: 400.0px;" /></a>
<p class="caption"><span class="caption-text">Homologous gene types.</span><a class="headerlink" href="#id35" title="Permalink to this image"></a></p>
<div class="legend">
<p>Homologs are formed from duplication events. Gene duplication events within a species result in paralogs (or paralogous genes). These can evolve distinct new functions (e.g. the globins). Speciation events lead leads to orthologs (or orthologous genes) whose function is putatively unchanged in the different species. The phylogeny between orthologs that remain unique in their respective genomes should reflect the species tree.</p>
<p>Figure from <span id="id1">[<a class="reference internal" href="#id23" title="David B. Searls. Pharmacophylogenomics: genes, evolution and drug targets. Nature Reviews Drug Discovery, 2:613–623, 2003. URL: https://doi.org/10.1038/nrd1152, doi:10.1038/nrd1152.">Sea03</a>]</span>.</p>
</div>
</div>
</div>
<div class="section" id="exercises">
<h2><span class="section-number">7.3.2. </span>Exercises<a class="headerlink" href="#exercises" title="Permalink to this headline"></a></h2>
<p>I have queried the Ensembl database for some haemoglobins from Human and Macaque. For each database record, I’ve used Ensembl’s definition of the “canonical transcript” (or “major” transcript). I’ve then obtained the CDS (coding sequence) for each record. I named the sequences {Species Common Name}-{a letter}. I then aligned the sequences.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">Human-j</td><td><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="T_dna">T</span></td></tr>
<tr><td class="label">Human-z</td><td><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">T</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="C_dna">C</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">A</span><span class="C_dna">.</span><span class="C_dna">C</span></td></tr>
<tr><td class="label">Rhesus-e</td><td><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">T</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="C_dna">C</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span></td></tr>
<tr><td class="label">Rhesus-x</td><td><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">.</span></td></tr>
</table>
<p><i>4 x 447 (truncated to 4 x 30) dna alignment</i></p>
</div></div></div>
</div>
<ol class="arabic simple">
<li><p>Which of the above are orthologs?</p></li>
<li><p>Which are paralogs?</p></li>
<li><p>For these sequences, how would you describe the divergence between orthologs compared to that between paralogs?</p></li>
</ol>
<hr class="docutils" />
<p class="rubric">Citations</p>
<div class="docutils container" id="id2">
<dl class="citation">
<dt class="label" id="id23"><span class="brackets"><a class="fn-backref" href="#id1">Sea03</a></span></dt>
<dd><p>David B. Searls. Pharmacophylogenomics: genes, evolution and drug targets. <em>Nature Reviews Drug Discovery</em>, 2:613–623, 2003. URL: <a class="reference external" href="https://doi.org/10.1038/nrd1152">https://doi.org/10.1038/nrd1152</a>, <a class="reference external" href="https://doi.org/10.1038/nrd1152">doi:10.1038/nrd1152</a>.</p>
</dd>
</dl>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="intro.html" class="btn btn-neutral float-left" title="7.1. What is molecular evolution?" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="neutral_theory.html" class="btn btn-neutral float-right" title="7.4. The neutral theory of molecular evolution" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2022, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>