<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>7.5. Mutation – the origin of genetic variation &mdash; Topics in Bioinformatics 2021.09.28 documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="7.6. Statistical methods for molecular evolutionary analyses" href="max_likelihood.html" />
    <link rel="prev" title="7.4. The neutral theory of molecular evolution" href="neutral_theory.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">7. Molecular Evolution</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="intro.html">7.1. What is molecular evolution?</a></li>
<li class="toctree-l2"><a class="reference internal" href="intro.html#what-is-phylogenetics">7.2. What is phylogenetics?</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html">7.3. Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="neutral_theory.html">7.4. The neutral theory of molecular evolution</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">7.5. Mutation – the origin of genetic variation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#types-of-mutation">7.5.1. Types of mutation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#point-mutations">7.5.2. Point mutations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#statistical-measures-of-sequence-composition-that-relate-to-mutation">7.5.3. Statistical measures of sequence composition that relate to mutation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="max_likelihood.html">7.6. Statistical methods for molecular evolutionary analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="substitution_models.html">7.7. Substitution models</a></li>
<li class="toctree-l2"><a class="reference internal" href="substitution_models.html#exercises">7.8. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="hypothesis_testing.html">7.9. Hypothesis Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="hypothesis_testing.html#exercises">7.10. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="glossary.html">7.11. Glossary of terms</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">9. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">10. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">11. Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html"><span class="section-number">7. </span>Molecular Evolution</a> &raquo;</li>
      <li><span class="section-number">7.5. </span>Mutation – the origin of genetic variation</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/molevol/mutation.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="mutation-the-origin-of-genetic-variation">
<h1><span class="section-number">7.5. </span>Mutation – the origin of genetic variation<a class="headerlink" href="#mutation-the-origin-of-genetic-variation" title="Permalink to this headline"></a></h1>
<p>Mutation manifests as a change in the primary DNA <a class="footnote-reference brackets" href="#id3" id="id1">1</a> sequence of an organism. Only mutations that occur within the cell lineage that produce sex cells, the germline, can have an influence on future generations. We refer to these as germline mutations.</p>
<div class="sidebar" id="mutant-fate">
<p class="sidebar-title">What happens to a mutation?</p>
<a class="reference internal image-reference" href="../_images/mutation_process.png"><img alt="../_images/mutation_process.png" src="../_images/mutation_process.png" style="width: 477.5px; height: 368.0px;" /></a>
<p>This process only applies to the germ cells.</p>
</div>
<p>In <a class="reference internal" href="#mutant-fate"><span class="std std-ref">the side bar</span></a>, we present a cartoon representation of the possible events that can affect a new genetic variant. The process begins at a single DNA molecule within a single cell as a lesion, a chemical change that disrupts the double-helix. If that lesion reverts to its original form or is correctly repaired by a DNA repair process, then we do not observed any change in the sequence. If this new variant is not transmitted to the next generation, then we never see it. If it does get transmitted, then it can be “lost” from the population purely by chance due to the process known as random genetic drift <a class="footnote-reference brackets" href="#id4" id="id2">2</a>. If the variant is deleterious, it may be lost as a consequence of natural selection eliminating the individuals that carry it, thus preventing them from contributing to the next generation. Finally, a variant that survives all of that can attain a frequency of 1.0 in a species. We apply the term substitution here since the original variant present in all members of a species has been substituted with this derived variant.</p>
<dl class="footnote brackets">
<dt class="label" id="id3"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Of course, some viruses use RNA for the genome, but we limit ourselves to DNA here.</p>
</dd>
<dt class="label" id="id4"><span class="brackets"><a class="fn-backref" href="#id2">2</a></span></dt>
<dd><p>If the variant survives to a sufficient population frequency, we refer to it as a polymorphism. Polymorphic means “many forms”. In terms of genetic variation, this is most often presented as SNPs (single nucleotide polymorphisms) or the largely synonymous term SNV (single nucleotide variant). The latter terms is probably better, since the original definition for polymorphism applied only to loci for which the minor allele frequency was <span class="math notranslate nohighlight">\(\ge 0.01\)</span> (and before that <span class="math notranslate nohighlight">\(\ge 0.05\)</span>). Since increase capacity to survey genetic variation, those arbitrary cutoffs have less value.</p>
</dd>
</dl>
<p>Any process that influences the different stages in a characteristic manner will contribute to the characteristics of the distribution of genetic variation.</p>
<div class="section" id="types-of-mutation">
<span id="index-0"></span><h2><span class="section-number">7.5.1. </span>Types of mutation<a class="headerlink" href="#types-of-mutation" title="Permalink to this headline"></a></h2>
<p>There are a 3 main mutation types.</p>
<dl class="simple">
<dt>Point mutations</dt><dd><p>The state of a single nucleotide changes between parent and daughter DNA sequence. The daughter DNA molecule is the same length as the parent.</p>
</dd>
<dt>Deletions</dt><dd><p>Deletions of DNA range in scale from single nucleotides to large genomic segments. The daughter DNA molecule is shorter than the parent.</p>
</dd>
<dt>Insertions</dt><dd><p>Range in scale from single to multiple nucleotides up to large genomic segments. The daughter DNA molecule is longer than the parent.</p>
</dd>
</dl>
<p>Within each of these, there are substantial sub-categories. We are interested in insertion / deletion events (indels) insomuch as they motivate the need for sequence alignment algorithms. Our principal focus, however, is on point mutation processes since those are the dominant type of genetic variation data.</p>
</div>
<div class="section" id="point-mutations">
<span id="id5"></span><h2><span class="section-number">7.5.2. </span>Point mutations<a class="headerlink" href="#point-mutations" title="Permalink to this headline"></a></h2>
<p id="index-1">There are 12 distinct point mutation events since each nucleotide can mutate to 3 possible alternates (e.g. A<span class="math notranslate nohighlight">\(\rightarrow\)</span>C, A<span class="math notranslate nohighlight">\(\rightarrow\)</span>G, A<span class="math notranslate nohighlight">\(\rightarrow\)</span>T). These are often categorised by the chemical classes of the bases involved. Specifically, transitions and transversions. Point mutations that start and end in bases belonging to the same chemical class are referred to as transitions, i.e. changes involving A<span class="math notranslate nohighlight">\(\rightarrow\)</span>G, G<span class="math notranslate nohighlight">\(\rightarrow\)</span>A, C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T, T<span class="math notranslate nohighlight">\(\rightarrow\)</span>C (blue lines in the <a class="reference internal" href="#directions"><span class="std std-ref">figure</span></a>). The remaining changes are assigned to the transversion category. As it turns out, there are differences in the rate at which mutations of these categories are observed and it has been argued that the excess of transitions reflects the chemical properties of DNA <span id="id6">[<a class="reference internal" href="#id38" title="M D Topal and J R Fresco. Complementary base pairing and the origin of substitution mutations. Nature, 263:285–289, 1976.">TF76</a>]</span>.</p>
<div class="sidebar" id="directions">
<p class="sidebar-title">Mutation directions</p>
<div class="graphviz"><object data="../_images/graphviz-b4cb238982177981577ef5d3c05d9d10e3724938.svg" type="image/svg+xml" class="graphviz">
<p class="warning">digraph point_mutants {
node [shape=none arrowhead=vee];
layout=circo

A -&gt; C [dir=both];
A -&gt; G [dir=both color=blue];
A -&gt; T [dir=both];
C -&gt; G [dir=both];
C -&gt; T [dir=both color=blue];
G -&gt; T [dir=both];
}</p></object></div>
<p>The different point mutations.</p>
<p>The blue lines indicate transition mutations, point mutations between bases that belong to the same chemical class.</p>
</div>
<p>But the dominance of transition mutations reflects more than just the intrinsic properties of the canonical bases. In vertebrates and many other organisms, the modified base 5-methyl-cytosine (hereafter <sup>5</sup>mC or methylated cytosine) is present. In vertebrates, at least, this modification can be used to encode information – switching between methylated and unmethylated states is associated with changes to gene expression of flanking genes. As such, <sup>5</sup>mC is a part of the epigenetic control layer. The modified base <sup>5</sup>mC is also hypermutable <span id="id7">[<a class="reference internal" href="#id37" title="C Coulondre, J H Miller, P J Farabaugh, and W Gilbert. Molecular basis of base substitution hotspots in \emp Escherichia coli. Nature, 274:775–780, 1978.">CMFG78</a>]</span>. The deamination of <sup>5</sup>mC (a hydrolysis reaction) occurs at a rate ~10x the same reaction of unmethylated C. The lesion arising from these reactions also differ, with <sup>5</sup>mC producing T while hydrolysis of unmethylated C produces uracil (U). These lesions cause a pairing mismatch in the helix, triggering DNA repair mechanisms. As you might expect, a repair system will do a better job of reverting U:G to the correct C:G compared with resolving a T:G mismatch.</p>
<p id="index-2">Where <sup>5</sup>mC mutagenesis gets even more interesting is that this is an enzymatically induced modification and the recognition sequence for the DNA methylase is a C followed by a G, denote CpG (the p stands for the phosphodiester bond between adjacent nucleotides). This sequence “context dependence” of the introduction of the base modification therefore results in a context dependence of C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T point mutations (see <a class="reference internal" href="#ctot-motif"><span class="std std-ref">Sidebar Figure</span></a>).</p>
<div class="sidebar" id="ctot-motif">
<p class="sidebar-title">Context dependence of C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T point mutations</p>
<div class="figure align-default" id="id53">
<img alt="../_images/CtoT-human-intergenic.svg" src="../_images/CtoT-human-intergenic.svg" /><p class="caption"><span class="caption-text">Information analysis of human intergenic SNPs resulting from a C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T point mutation <span id="id8">[<a class="reference internal" href="../seqcomp/information.html#id31" title="Yicheng Zhu, Teresa Neeman, Von Bing Yap, and Gavin A Huttley. Statistical methods for identifying sequence motifs affecting point mutations. Genetics, 205:843-856, 2017. doi:10.1534/genetics.116.195677.">ZNYH17</a>]</span>.</span><a class="headerlink" href="#id53" title="Permalink to this image"></a></p>
<div class="legend">
<p>RE is relative entropy. Position is relative to the point mutation (at 0). The normal letter orientation in the plot indicates that base was over-represented in mutant sequences compared to the reference distribution. The rotated orientation indicates that base was under-represented in mutant sequences.</p>
</div>
</div>
</div>
</div>
<div class="section" id="statistical-measures-of-sequence-composition-that-relate-to-mutation">
<h2><span class="section-number">7.5.3. </span>Statistical measures of sequence composition that relate to mutation<a class="headerlink" href="#statistical-measures-of-sequence-composition-that-relate-to-mutation" title="Permalink to this headline"></a></h2>
<p>As the C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T case illustrates, chemical and metabolic processes affect how mutation occurs. To further illustrate this, we consider an additional property of DNA sequences – strand.</p>
<p>When we discuss processes via which lesions form in DNA, we are predominantly referencing chemical reactions affecting the base part of a nucleotide. Thymidine dimers arise from UV light induced covalent bonds between Thymine bases that are physically adjacent <em>on the same DNA strand</em>. This strand orientation leads to a simple question: Does mutation occur in a strand symmetric way?</p>
<p>To address this, let’s think back to what we actually observe. We do not observe the mutation process, we observe the outcome <a class="footnote-reference brackets" href="#id13" id="id9">3</a>. Let’s assume we detect a G<span class="math notranslate nohighlight">\(\rightarrow\)</span>A difference between the parent DNA sequence and its immediate descendant. We represent DNA sequences by picking one strand and displaying that information only <a class="footnote-reference brackets" href="#id14" id="id10">4</a>, often an entirely arbitrary choice. Accordingly, the designation of mutation direction is also arbitrary and, for our example, its possible this mutation was in fact a C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T on the other strand. If the mutation was of CpG→CpA, it’s likely the actual mutation involved the <sup>5</sup>mC on the opposite strand since CpG is a strand symmetric dinucleotide (the reverse complement is also CpG).</p>
<p>Let’s consider a though experiment in where we run a mutagenesis experiment for a very long time on DNA that does not encode any information. In the absence of any biochemical biases, we expect mutation processes to occur with equal probability on the two strands. As a consequence, we expect at chemical equilibrium, the bases that form the canonical Watson-Crick base pairs to have equals counts on the strand, i.e. they are strand-symmetric. For instance, the DNA sequence ATGC is strand symmetric, as is AATTGC. The following “Skewness” statistics are used to quantify strand symmetry (or strand parity).</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}S_{AT} = \frac{A-T}{A+T}\\S_{GC} = \frac{G-C}{G+C}\end{aligned}\end{align} \]</div>
<p>These divide the difference in the counts of the Watson-Crick pairs by their total. If sequences are mutation has predominantly operated in a strand-symmetric manner throughout time, the expected value of both <span class="math notranslate nohighlight">\(S_{AT}\)</span> and <span class="math notranslate nohighlight">\(S_{GA}\)</span> is 0 <a class="footnote-reference brackets" href="#id15" id="id11">5</a>.</p>
<p>We present two figures from published work that prove strikingly informative. The <a class="reference internal" href="#dna-rep"><span class="std std-ref">first</span></a> concerns the putative influence of initiating DNA replication from a fixed location. It is conjectured that the distinct nature of DNA synthesis on leading versus lagging strands drives the appearance of striking asymmetries in some bacterial genomes <span id="id12">[<a class="reference internal" href="#id36" title="Jan Mràzek and Samuel Karlin. Strand compositional asymmetry in bacterial and large viral genomes. Proceedings of the National Academy of Sciences of the United States of America, 95:3720–3725, 1998. URL: www.pnas.org., doi:10.1073/pnas.95.7.3720.">MrazekK98</a>]</span>.</p>
<dl class="footnote brackets">
<dt class="label" id="id13"><span class="brackets"><a class="fn-backref" href="#id9">3</a></span></dt>
<dd><p>Except in specific experimental contexts.</p>
</dd>
<dt class="label" id="id14"><span class="brackets"><a class="fn-backref" href="#id10">4</a></span></dt>
<dd><p>Because of the Watson-Crick base-pairing rules, the other strand can be deduced and thus presenting it is redundant.</p>
</dd>
<dt class="label" id="id15"><span class="brackets"><a class="fn-backref" href="#id11">5</a></span></dt>
<dd><p>The order of the base counts in the statistics can differ between publications.</p>
</dd>
</dl>
<div class="sidebar" id="dna-rep">
<p class="sidebar-title">The influence of DNA replication</p>
<div class="figure align-default" id="id54">
<img alt="../_images/bsubtilits_symmetry.png" src="../_images/bsubtilits_symmetry.png" />
<p class="caption"><span class="caption-text"><em>Bacillus subtilis</em> genome.</span><a class="headerlink" href="#id54" title="Permalink to this image"></a></p>
</div>
<div class="figure align-default" id="id55">
<img alt="../_images/synechocystis_symmetry.png" src="../_images/synechocystis_symmetry.png" />
<p class="caption"><span class="caption-text"><em>Synechocystis</em> PCC6803 genome.</span><a class="headerlink" href="#id55" title="Permalink to this image"></a></p>
</div>
<p>Panels copied from Figure 1 of <span id="id16">[<a class="reference internal" href="#id36" title="Jan Mràzek and Samuel Karlin. Strand compositional asymmetry in bacterial and large viral genomes. Proceedings of the National Academy of Sciences of the United States of America, 95:3720–3725, 1998. URL: www.pnas.org., doi:10.1073/pnas.95.7.3720.">MrazekK98</a>]</span>. The <span class="math notranslate nohighlight">\(y\)</span>-axis is <span class="math notranslate nohighlight">\(-S_{GC}\)</span> computed from a 50kb sliding window across the corresponding genome. The statistic is assigned to the middle base of thew window. The arrow indicates the origin of replication.</p>
</div>
<p>The second example concerns the distribution of strand symmetry around genes in humans <span id="id17">[<a class="reference internal" href="#id35" title="M Touchon, S Nicolay, A Arneodo, Y D'Aubenton-Carafa, and C Thermes. Transcription-coupled TA and GC strand asymmetries in the human genome. FEBS Lett, 555:579–582, 2003.">TNA+03</a>]</span>. In this case, the proposed biochemical mechanism is transcription couple DNA repair. In simplistic terms, this is a DNA damage repair system that is induced by a stalled RNA polymerase. The repair has been shown to be limited to the transcribed strand. This observation implies that the non-transcribed strand receives less scrutiny by lesion repair processes. This asymmetry also manifests in the SNPs that are present in humans today, indicating the influence is active <span id="id18">[<a class="reference internal" href="#id34" title="Helmut Simon and Gavin Huttley. Quantifying Influences on Intragenomic Mutation Rate. G3: Genes|Genomes|Genetics, 10:g3.401335.2020, 2020. URL: https://doi.org/10.1534/g3.120.401335, doi:10.1534/g3.120.401335.">SH20</a>]</span>.</p>
<div class="sidebar">
<p class="sidebar-title">Transcription associated mutation asymmetries</p>
<div class="figure align-default">
<img alt="../_images/strand_skew.png" src="../_images/strand_skew.png" />
</div>
<p>Statistics were calculated using the human genome sequence in 1kb windows around genes. The left column shows the transcriptional start site (TSS) at index 0. The <span class="math notranslate nohighlight">\(x\)</span>-axis values correspond to distances to the TSS in the left column. In the right column they correspond to the distance from the 3`-terminus of the annotated gene transcript. The <span class="math notranslate nohighlight">\(y\)</span>-axis values are the mean skew statistic for that position from all human genes.</p>
<p>Copied from Figure 3 of <span id="id19">[<a class="reference internal" href="#id35" title="M Touchon, S Nicolay, A Arneodo, Y D'Aubenton-Carafa, and C Thermes. Transcription-coupled TA and GC strand asymmetries in the human genome. FEBS Lett, 555:579–582, 2003.">TNA+03</a>]</span>.</p>
</div>
<hr class="docutils" />
<p class="rubric">Citations</p>
<p id="id20"><dl class="citation">
<dt class="label" id="id37"><span class="brackets"><a class="fn-backref" href="#id7">CMFG78</a></span></dt>
<dd><p>C Coulondre, J H Miller, P J Farabaugh, and W Gilbert. Molecular basis of base substitution hotspots in \emp Escherichia coli. <em>Nature</em>, 274:775–780, 1978.</p>
</dd>
<dt class="label" id="id36"><span class="brackets">MrazekK98</span><span class="fn-backref">(<a href="#id12">1</a>,<a href="#id16">2</a>)</span></dt>
<dd><p>Jan Mràzek and Samuel Karlin. Strand compositional asymmetry in bacterial and large viral genomes. <em>Proceedings of the National Academy of Sciences of the United States of America</em>, 95:3720–3725, 1998. URL: <a class="reference external" href="www.pnas.org.">www.pnas.org.</a>, <a class="reference external" href="https://doi.org/10.1073/pnas.95.7.3720">doi:10.1073/pnas.95.7.3720</a>.</p>
</dd>
<dt class="label" id="id34"><span class="brackets"><a class="fn-backref" href="#id18">SH20</a></span></dt>
<dd><p>Helmut Simon and Gavin Huttley. Quantifying Influences on Intragenomic Mutation Rate. <em>G3: Genes|Genomes|Genetics</em>, 10:g3.401335.2020, 2020. URL: <a class="reference external" href="https://doi.org/10.1534/g3.120.401335">https://doi.org/10.1534/g3.120.401335</a>, <a class="reference external" href="https://doi.org/10.1534/g3.120.401335">doi:10.1534/g3.120.401335</a>.</p>
</dd>
<dt class="label" id="id38"><span class="brackets"><a class="fn-backref" href="#id6">TF76</a></span></dt>
<dd><p>M D Topal and J R Fresco. Complementary base pairing and the origin of substitution mutations. <em>Nature</em>, 263:285–289, 1976.</p>
</dd>
<dt class="label" id="id35"><span class="brackets">TNA+03</span><span class="fn-backref">(<a href="#id17">1</a>,<a href="#id19">2</a>)</span></dt>
<dd><p>M Touchon, S Nicolay, A Arneodo, Y D'Aubenton-Carafa, and C Thermes. Transcription-coupled TA and GC strand asymmetries in the human genome. <em>FEBS Lett</em>, 555:579–582, 2003.</p>
</dd>
<dt class="label" id="id46"><span class="brackets"><a class="fn-backref" href="#id8">ZNYH17</a></span></dt>
<dd><p>Yicheng Zhu, Teresa Neeman, Von Bing Yap, and Gavin A Huttley. Statistical methods for identifying sequence motifs affecting point mutations. <em>Genetics</em>, 205:843–856, 2017. <a class="reference external" href="https://doi.org/10.1534/genetics.116.195677">doi:10.1534/genetics.116.195677</a>.</p>
</dd>
</dl>
</p>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="neutral_theory.html" class="btn btn-neutral float-left" title="7.4. The neutral theory of molecular evolution" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="max_likelihood.html" class="btn btn-neutral float-right" title="7.6. Statistical methods for molecular evolutionary analyses" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2021, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>