
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>7.5. Mutation – the origin of genetic variation &#8212; Topics in Bioinformatics</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=fd3f3429" />
    <link rel="stylesheet" type="text/css" href="../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css?v=3ce756a4" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=9bad7206"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/thebelab-helper.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'molevol/mutation';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="7.6. Statistical methods for molecular evolutionary analyses" href="max_likelihood.html" />
    <link rel="prev" title="7.4. The neutral theory of molecular evolution" href="neutral_theory.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Topics in Bioinformatics</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">7. Molecular Evolution</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="intro.html">7.1. What is molecular evolution?</a></li>

<li class="toctree-l2"><a class="reference internal" href="data.html">7.3. Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="neutral_theory.html">7.4. The neutral theory of molecular evolution</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">7.5. Mutation – the origin of genetic variation</a></li>
<li class="toctree-l2"><a class="reference internal" href="max_likelihood.html">7.6. Statistical methods for molecular evolutionary analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="substitution_models.html">7.7. Substitution models</a></li>

<li class="toctree-l2"><a class="reference internal" href="hypothesis_testing.html">7.9. Hypothesis Testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="glossary.html">7.11. Glossary of terms</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../statistics/index.html">9. Statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/edit/main/molevol/mutation.rst" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/issues/new?title=Issue%20on%20page%20%2Fmolevol/mutation.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/molevol/mutation.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Mutation – the origin of genetic variation</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-mutation">7.5.1. Types of mutation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#point-mutations">7.5.2. Point mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistical-measures-of-sequence-composition-that-relate-to-mutation">7.5.3. Statistical measures of sequence composition that relate to mutation</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="mutation-the-origin-of-genetic-variation">
<h1><span class="section-number">7.5. </span>Mutation – the origin of genetic variation<a class="headerlink" href="#mutation-the-origin-of-genetic-variation" title="Link to this heading">#</a></h1>
<p>Mutation manifests as a change in the primary DNA <a class="footnote-reference brackets" href="#id3" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a> sequence of an organism. Only mutations that occur within the cell lineage that produce sex cells, the germline, can have an influence on future generations. We refer to these as germline mutations.</p>
<aside class="margin sidebar" id="mutant-fate">
<p class="sidebar-title">What happens to a mutation?</p>
<a class="reference internal image-reference" href="../_images/mutation_process.png"><img alt="../_images/mutation_process.png" src="../_images/mutation_process.png" style="width: 477.5px; height: 368.0px;" />
</a>
<p>This process only applies to the germ cells.</p>
</aside>
<p>In <a class="reference internal" href="#mutant-fate"><span class="std std-ref">the side bar</span></a>, we present a cartoon representation of the possible events that can affect a new genetic variant. The process begins at a single DNA molecule within a single cell as a lesion, a chemical change that disrupts the double-helix. If that lesion reverts to its original form or is correctly repaired by a DNA repair process, then we do not observed any change in the sequence. If this new variant is not transmitted to the next generation, then we never see it. If it does get transmitted, then it can be “lost” from the population purely by chance due to the process known as random genetic drift <a class="footnote-reference brackets" href="#id4" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. If the variant is deleterious, it may be lost as a consequence of natural selection eliminating the individuals that carry it, thus preventing them from contributing to the next generation. Finally, a variant that survives all of that can attain a frequency of 1.0 in a species. We apply the term substitution here since the original variant present in all members of a species has been substituted with this derived variant.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id3" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>Of course, some viruses use RNA for the genome, but we limit ourselves to DNA here.</p>
</aside>
<aside class="footnote brackets" id="id4" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">2</a><span class="fn-bracket">]</span></span>
<p>If the variant survives to a sufficient population frequency, we refer to it as a polymorphism. Polymorphic means “many forms”. In terms of genetic variation, this is most often presented as SNPs (single nucleotide polymorphisms) or the largely synonymous term SNV (single nucleotide variant). The latter terms is probably better, since the original definition for polymorphism applied only to loci for which the minor allele frequency was <span class="math notranslate nohighlight">\(\ge 0.01\)</span> (and before that <span class="math notranslate nohighlight">\(\ge 0.05\)</span>). Since increase capacity to survey genetic variation, those arbitrary cutoffs have less value.</p>
</aside>
</aside>
</aside>
<p>Any process that influences the different stages in a characteristic manner will contribute to the characteristics of the distribution of genetic variation.</p>
<section id="types-of-mutation">
<span id="index-0"></span><h2><span class="section-number">7.5.1. </span>Types of mutation<a class="headerlink" href="#types-of-mutation" title="Link to this heading">#</a></h2>
<p>There are a 3 main mutation types.</p>
<dl class="simple">
<dt>Point mutations</dt><dd><p>The state of a single nucleotide changes between parent and daughter DNA sequence. The daughter DNA molecule is the same length as the parent.</p>
</dd>
<dt>Deletions</dt><dd><p>Deletions of DNA range in scale from single nucleotides to large genomic segments. The daughter DNA molecule is shorter than the parent.</p>
</dd>
<dt>Insertions</dt><dd><p>Range in scale from single to multiple nucleotides up to large genomic segments. The daughter DNA molecule is longer than the parent.</p>
</dd>
</dl>
<p>Within each of these, there are substantial sub-categories. We are interested in insertion / deletion events (indels) insomuch as they motivate the need for sequence alignment algorithms. Our principal focus, however, is on point mutation processes since those are the dominant type of genetic variation data.</p>
</section>
<section id="point-mutations">
<span id="id5"></span><h2><span class="section-number">7.5.2. </span>Point mutations<a class="headerlink" href="#point-mutations" title="Link to this heading">#</a></h2>
<p id="index-1">There are 12 distinct point mutation events since each nucleotide can mutate to 3 possible alternates (e.g. A<span class="math notranslate nohighlight">\(\rightarrow\)</span>C, A<span class="math notranslate nohighlight">\(\rightarrow\)</span>G, A<span class="math notranslate nohighlight">\(\rightarrow\)</span>T). These are often categorised by the chemical classes of the bases involved. Specifically, transitions and transversions. Point mutations that start and end in bases belonging to the same chemical class are referred to as transitions, i.e. changes involving A<span class="math notranslate nohighlight">\(\rightarrow\)</span>G, G<span class="math notranslate nohighlight">\(\rightarrow\)</span>A, C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T, T<span class="math notranslate nohighlight">\(\rightarrow\)</span>C (blue lines in the <a class="reference internal" href="#directions"><span class="std std-ref">figure</span></a>). The remaining changes are assigned to the transversion category. As it turns out, there are differences in the rate at which mutations of these categories are observed, and it has been argued that the excess of transitions reflects the chemical properties of DNA <span id="id6">[<a class="reference internal" href="#id38" title="M D Topal and J R Fresco. Complementary base pairing and the origin of substitution mutations. Nature, 263:285–289, 1976.">TF76</a>]</span>.</p>
<aside class="margin sidebar" id="directions">
<p class="sidebar-title">Mutation directions</p>
<div class="graphviz"><object data="../_images/graphviz-b4cb238982177981577ef5d3c05d9d10e3724938.svg" type="image/svg+xml" class="graphviz">
<p class="warning">digraph point_mutants {
node [shape=none arrowhead=vee];
layout=circo

A -&gt; C [dir=both];
A -&gt; G [dir=both color=blue];
A -&gt; T [dir=both];
C -&gt; G [dir=both];
C -&gt; T [dir=both color=blue];
G -&gt; T [dir=both];
}</p></object></div>
<p>The different point mutations.</p>
<p>The blue lines indicate transition mutations, point mutations between bases that belong to the same chemical class – between the purines (A/G) or pyrimidines (C/T).</p>
</aside>
<p>The dominance of transition mutations reflects more than just the intrinsic properties of the canonical bases. The modified base 5-methyl-cytosine (hereafter <sup>5</sup>mC or methylated cytosine) is present in vertebrates and many other organisms. In vertebrates, at least, this modification can be used to encode information – switching between methylated and unmethylated states is associated with changes to gene expression of flanking genes. As such, <sup>5</sup>mC is a part of the epigenetic control layer. The modified base <sup>5</sup>mC is also hypermutable <span id="id7">[<a class="reference internal" href="#id37" title="C Coulondre, J H Miller, P J Farabaugh, and W Gilbert. Molecular basis of base substitution hotspots in \emp Escherichia coli. Nature, 274:775–780, 1978.">CMFG78</a>]</span>. The deamination of <sup>5</sup>mC (a hydrolysis reaction) occurs at a rate ~10x the same reaction of unmethylated C. The lesion arising from these reactions also differ, with <sup>5</sup>mC producing T while hydrolysis of unmethylated C produces uracil (U). These lesions cause a pairing mismatch in the helix, triggering DNA repair mechanisms. As U is typically seen in RNA we might reasonably expect a repair system will do a better job of reverting U:G to the correct C:G compared with resolving a T:G mismatch.</p>
<p id="index-2">Where <sup>5</sup>mC mutagenesis gets even more interesting is that this is an enzymatically induced modification and the recognition sequence for the DNA methylase is a C followed by a G, denote CpG (the p stands for the phosphodiester bond between adjacent nucleotides). This sequence “context dependent” introduction of the base modification therefore results in a context dependence of C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T point mutations (see <a class="reference internal" href="#ctot-motif"><span class="std std-ref">Sidebar Figure</span></a>).</p>
<aside class="margin sidebar" id="ctot-motif">
<p class="sidebar-title">Context dependence of C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T point mutations</p>
<figure class="align-default">
<img alt="../_images/CtoT-human-intergenic.svg" src="../_images/CtoT-human-intergenic.svg" /></figure>
<p>Information analysis of human intergenic SNPs resulting from a C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T point mutation <span id="id8">[<a class="reference internal" href="../seqcomp/information.html#id31" title="Yicheng Zhu, Teresa Neeman, Von Bing Yap, and Gavin A Huttley. Statistical methods for identifying sequence motifs affecting point mutations. Genetics, 205:843-856, 2017. doi:10.1534/genetics.116.195677.">ZNYH17</a>]</span>.</p>
<p>RE is relative entropy. Position is relative to the point mutation (at 0). The normal letter orientation in the plot indicates that base was over-represented in mutant sequences compared to the reference distribution. The rotated orientation indicates that base was under-represented in mutant sequences.</p>
</aside>
</section>
<section id="statistical-measures-of-sequence-composition-that-relate-to-mutation">
<h2><span class="section-number">7.5.3. </span>Statistical measures of sequence composition that relate to mutation<a class="headerlink" href="#statistical-measures-of-sequence-composition-that-relate-to-mutation" title="Link to this heading">#</a></h2>
<p>As the C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T case illustrates, chemical and metabolic processes affect how mutation occurs. To further illustrate this, we consider an additional property of DNA sequences – strand.</p>
<p>When we discuss processes via which lesions form in DNA, we are predominantly referencing chemical reactions affecting the base part of a nucleotide. Thymidine dimers arise from UV light induced covalent bonds between Thymine bases that are physically adjacent <em>on the same DNA strand</em>. This strand orientation leads to a simple question: Does mutation occur in a strand symmetric way?</p>
<p>To address this, let’s think back to what we actually observe. We do not observe the mutation process, we observe the outcome <a class="footnote-reference brackets" href="#id13" id="id9" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a>. Let’s assume we detect a G<span class="math notranslate nohighlight">\(\rightarrow\)</span>A difference between the parent DNA sequence and its immediate descendant. We represent DNA sequences by picking one strand and displaying that information only <a class="footnote-reference brackets" href="#id14" id="id10" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>, often an entirely arbitrary choice. Accordingly, the designation of mutation direction is also arbitrary and, for our example, its possible this mutation was in fact a C<span class="math notranslate nohighlight">\(\rightarrow\)</span>T on the other strand. If the mutation was of CpG→CpA, it’s likely the actual mutation involved the <sup>5</sup>mC on the opposite strand since CpG is a strand symmetric dinucleotide (the reverse complement is also CpG).</p>
<p>Let’s consider a though experiment in where we run a mutagenesis experiment for a very long time on DNA that does not encode any information. In the absence of any biochemical biases, we expect mutation processes to occur with equal probability on the two strands. As a consequence, we expect at chemical equilibrium, the bases that form the canonical Watson-Crick base pairs to have equals counts on the strand, i.e. they are strand-symmetric. For instance, the DNA sequence ATGC is strand symmetric, as is AATTGC. The following “Skewness” statistics are used to quantify strand symmetry (or strand parity).</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}S_{AT} = \frac{A-T}{A+T}\\S_{GC} = \frac{G-C}{G+C}\end{aligned}\end{align} \]</div>
<p>These divide the difference in the counts of the Watson-Crick pairs by their total. If sequence mutation has predominantly operated in a strand-symmetric manner throughout time, the expected value of both <span class="math notranslate nohighlight">\(S_{AT}\)</span> and <span class="math notranslate nohighlight">\(S_{GA}\)</span> is 0 <a class="footnote-reference brackets" href="#id15" id="id11" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>.</p>
<p>We present two figures from published work that prove strikingly informative. The <a class="reference internal" href="#dna-rep"><span class="std std-ref">first</span></a> concerns the putative influence of initiating DNA replication from a fixed location. It is conjectured that the distinct nature of DNA synthesis on leading versus lagging strands drives the appearance of striking asymmetries in some bacterial genomes <span id="id12">[<a class="reference internal" href="#id36" title="Jan Mràzek and Samuel Karlin. Strand compositional asymmetry in bacterial and large viral genomes. Proceedings of the National Academy of Sciences of the United States of America, 95:3720–3725, 1998. URL: www.pnas.org., doi:10.1073/pnas.95.7.3720.">MrazekK98</a>]</span>.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id13" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">3</a><span class="fn-bracket">]</span></span>
<p>Except in specific experimental contexts.</p>
</aside>
<aside class="footnote brackets" id="id14" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id10">4</a><span class="fn-bracket">]</span></span>
<p>Because of the Watson-Crick base-pairing rules, the other strand can be deduced and thus presenting it is redundant.</p>
</aside>
<aside class="footnote brackets" id="id15" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">5</a><span class="fn-bracket">]</span></span>
<p>The order of the base counts in the statistics can differ between publications.</p>
</aside>
</aside>
</aside>
<aside class="margin sidebar" id="dna-rep">
<p class="sidebar-title">The influence of DNA replication</p>
<figure class="align-default" id="id56">
<img alt="../_images/bsubtilits_symmetry.png" src="../_images/bsubtilits_symmetry.png" />
<figcaption>
<p><span class="caption-text"><em>Bacillus subtilis</em> genome.</span><a class="headerlink" href="#id56" title="Link to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="id57">
<img alt="../_images/synechocystis_symmetry.png" src="../_images/synechocystis_symmetry.png" />
<figcaption>
<p><span class="caption-text"><em>Synechocystis</em> PCC6803 genome.</span><a class="headerlink" href="#id57" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Panels copied from Figure 1 of <span id="id16">[<a class="reference internal" href="#id36" title="Jan Mràzek and Samuel Karlin. Strand compositional asymmetry in bacterial and large viral genomes. Proceedings of the National Academy of Sciences of the United States of America, 95:3720–3725, 1998. URL: www.pnas.org., doi:10.1073/pnas.95.7.3720.">MrazekK98</a>]</span>. The <span class="math notranslate nohighlight">\(y\)</span>-axis is <span class="math notranslate nohighlight">\(-S_{GC}\)</span> computed from a 50kb sliding window across the corresponding genome. The statistic is assigned to the middle base of thew window. The arrow indicates the origin of replication.</p>
</aside>
<p>The second example concerns the distribution of strand symmetry around genes in humans <span id="id17">[<a class="reference internal" href="#id35" title="M Touchon, S Nicolay, A Arneodo, Y D'Aubenton-Carafa, and C Thermes. Transcription-coupled TA and GC strand asymmetries in the human genome. FEBS Lett, 555:579–582, 2003.">TNA+03</a>]</span>. In this case, the proposed biochemical mechanism is transcription coupled DNA repair. In simplistic terms, this is a DNA damage repair system that is induced by a stalled RNA polymerase. The repair has been shown to be limited to the transcribed strand. This observation implies that the non-transcribed strand receives less scrutiny by lesion repair processes. This asymmetry also manifests in the SNPs that are present in humans today, indicating the influence is active <span id="id18">[<a class="reference internal" href="#id34" title="Helmut Simon and Gavin Huttley. Quantifying Influences on Intragenomic Mutation Rate. G3: Genes|Genomes|Genetics, 10:g3.401335.2020, 2020. URL: https://doi.org/10.1534/g3.120.401335, doi:10.1534/g3.120.401335.">SH20</a>]</span>.</p>
<aside class="margin sidebar">
<p class="sidebar-title">Transcription associated mutation asymmetries</p>
<figure class="align-default">
<img alt="../_images/strand_skew.png" src="../_images/strand_skew.png" />
</figure>
<p>Statistics were calculated using the human genome sequence in 1kb windows around genes. The left column shows the transcriptional start site (TSS) at index 0. The <span class="math notranslate nohighlight">\(x\)</span>-axis values correspond to distances to the TSS in the left column. In the right column they correspond to the distance from the 3`-terminus of the annotated gene transcript. The <span class="math notranslate nohighlight">\(y\)</span>-axis values are the mean skew statistic for that position from all human genes.</p>
<p>Copied from Figure 3 of <span id="id19">[<a class="reference internal" href="#id35" title="M Touchon, S Nicolay, A Arneodo, Y D'Aubenton-Carafa, and C Thermes. Transcription-coupled TA and GC strand asymmetries in the human genome. FEBS Lett, 555:579–582, 2003.">TNA+03</a>]</span>.</p>
</aside>
<hr class="docutils" />
<p class="rubric">Citations</p>
<div class="docutils container" id="id20">
<div role="list" class="citation-list">
<div class="citation" id="id37" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">CMFG78</a><span class="fn-bracket">]</span></span>
<p>C Coulondre, J H Miller, P J Farabaugh, and W Gilbert. Molecular basis of base substitution hotspots in \emp Escherichia coli. <em>Nature</em>, 274:775–780, 1978.</p>
</div>
<div class="citation" id="id36" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>MrazekK98<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id12">1</a>,<a role="doc-backlink" href="#id16">2</a>)</span>
<p>Jan Mràzek and Samuel Karlin. Strand compositional asymmetry in bacterial and large viral genomes. <em>Proceedings of the National Academy of Sciences of the United States of America</em>, 95:3720–3725, 1998. URL: <a class="reference external" href="www.pnas.org.">www.pnas.org.</a>, <a class="reference external" href="https://doi.org/10.1073/pnas.95.7.3720">doi:10.1073/pnas.95.7.3720</a>.</p>
</div>
<div class="citation" id="id34" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id18">SH20</a><span class="fn-bracket">]</span></span>
<p>Helmut Simon and Gavin Huttley. Quantifying Influences on Intragenomic Mutation Rate. <em>G3: Genes|Genomes|Genetics</em>, 10:g3.401335.2020, 2020. URL: <a class="reference external" href="https://doi.org/10.1534/g3.120.401335">https://doi.org/10.1534/g3.120.401335</a>, <a class="reference external" href="https://doi.org/10.1534/g3.120.401335">doi:10.1534/g3.120.401335</a>.</p>
</div>
<div class="citation" id="id38" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">TF76</a><span class="fn-bracket">]</span></span>
<p>M D Topal and J R Fresco. Complementary base pairing and the origin of substitution mutations. <em>Nature</em>, 263:285–289, 1976.</p>
</div>
<div class="citation" id="id35" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>TNA+03<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id17">1</a>,<a role="doc-backlink" href="#id19">2</a>)</span>
<p>M Touchon, S Nicolay, A Arneodo, Y D'Aubenton-Carafa, and C Thermes. Transcription-coupled TA and GC strand asymmetries in the human genome. <em>FEBS Lett</em>, 555:579–582, 2003.</p>
</div>
<div class="citation" id="id46" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">ZNYH17</a><span class="fn-bracket">]</span></span>
<p>Yicheng Zhu, Teresa Neeman, Von Bing Yap, and Gavin A Huttley. Statistical methods for identifying sequence motifs affecting point mutations. <em>Genetics</em>, 205:843–856, 2017. <a class="reference external" href="https://doi.org/10.1534/genetics.116.195677">doi:10.1534/genetics.116.195677</a>.</p>
</div>
</div>
</div>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="neutral_theory.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">7.4. </span>The neutral theory of molecular evolution</p>
      </div>
    </a>
    <a class="right-next"
       href="max_likelihood.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">7.6. </span>Statistical methods for molecular evolutionary analyses</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-mutation">7.5.1. Types of mutation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#point-mutations">7.5.2. Point mutations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#statistical-measures-of-sequence-composition-that-relate-to-mutation">7.5.3. Statistical measures of sequence composition that relate to mutation</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gavin Huttley
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2020-2024, Gavin Huttley.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>