<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>2. Prologue &mdash; Topics in Bioinformatics 2022.07.27 documentation</title><link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
      <link rel="stylesheet" href="_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
      <link rel="stylesheet" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="3. How to Contribute" href="contribute.html" />
    <link rel="prev" title="1. Acknowledgements" href="acknowledgements.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="version_control/index.html">9. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="testing/index.html">10. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">11. Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">2. </span>Prologue</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/intro.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="prologue">
<h1><span class="section-number">2. </span>Prologue<a class="headerlink" href="#prologue" title="Permalink to this headline"></a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>How is life encoded in nucleic acids? What is the minimum number of nucleotides required to encode a virus? Do all genetic differences between individuals contribute to phenotypic differences between them? What genetic change(s) causes humans to have unique cognitive abilities compared to our nearest great ape relatives? How many different types of microorganisms live in our gut and what role do they play in our biology? If you had <span class="math notranslate nohighlight">\(\sim 10^{10}\)</span> 250bp long DNA sequences that were randomly sampled from a single Human, could you assemble that individuals genome from scratch? If you sequence DNA extracted from soil, can you identify what species the DNA comes from? If you have an assembled genome of a novel species, can you identify where the genes in the genome are?</p>
<div class="sidebar">
<p class="sidebar-title">What’s the difference between using <a class="reference external" href="https://python.org">Python</a> versus using <a class="reference external" href="https://www.r-project.org/">R</a>?</p>
<p>Both languages provide very sophisticated capabilities for scientists, hence they are the dominant languages employed in Data Science. Both of these languages are also <a class="reference external" href="https://en.wikipedia.org/wiki/Open_source">Open Source</a> languages, developed by a community of volunteers <a class="footnote-reference brackets" href="#id7" id="id1">3</a>.</p>
<p>The major difference between <a class="reference external" href="https://python.org">Python</a> and <a class="reference external" href="https://www.r-project.org/">R</a> comes in how they are typically used <a class="footnote-reference brackets" href="#id8" id="id2">4</a>. Many of the concepts you will learn in the <a class="reference external" href="https://python.org">Python</a> topic – conditionals, iteration, functions – are not directly coded by most users of <a class="reference external" href="https://www.r-project.org/">R</a>. They are implicitly being used, but are “hidden” from users.</p>
<p>The way you will learn <a class="reference external" href="https://python.org">Python</a> is as a more “conventional” programming approach.</p>
</div>
<p>We do not know the answer to many of these questions. Despite that, and despite the fact that some of them are strictly concerned with biological phenomena, I am confident that they all require computers to attain a solution. I will go even further and claim that any question in the biological sciences we can come up with will require the use of computers to tackle it. Based on this, I think it is reasonable to make the general claim that biologists have a near absolute reliance on computers for undertaking their research <a class="footnote-reference brackets" href="#id5" id="id3">1</a>. But we can make that statement more specific: Biologists have a near absolute reliance on <strong>software</strong> to undertake their research <a class="footnote-reference brackets" href="#id6" id="id4">2</a>.</p>
<p>Just about every part of a biology wet-lab has some type of digital device, from pipette’s and scales to centrifuges. From confocal microscopes to scanning tunnelling electron microscopes. The more transformed data becomes from the natural system, the greater the involvement of software (and our reliance on it) to control those devices. But science is an enterprise whose objective is much more than simply the accumulation of data. Similarly, software can be much more than instructions for controlling the capture of data.</p>
<p>Science is about producing general models of how the natural world works. Good scientific models are powerful because they allow us to make predictions about what we will observe in circumstances that have not yet been encountered. Computer programs are one way in which we represent those models. They provide us with the means to evaluate, on a large scale, how well our model matches actual data. In other words, software is integral to scientific practice.</p>
<dl class="footnote brackets">
<dt class="label" id="id5"><span class="brackets"><a class="fn-backref" href="#id3">1</a></span></dt>
<dd><p>I suspect this holds true for all the sciences.</p>
</dd>
<dt class="label" id="id6"><span class="brackets"><a class="fn-backref" href="#id4">2</a></span></dt>
<dd><p>Software is just another word for computer program. The computer is just a piece of hardware but the software is the specific set of instructions that dictate what that hardware does.</p>
</dd>
</dl>
<p>The above is a big picture view of the role of software in science, but where does this course fit in? I can say with absolute certainty that we will not be answering most of the questions raised above (Doh!). What <em>you</em> will be doing is acquiring the skills necessary to understand what the analyses described above are, how to perform them and whether <em>you</em> should believe the results. <em>You</em> will develop the skills to write custom software whose output you can trust. In your future career, that software may be as simple as handling data formats, or completely novel models of biological systems. In all cases, your software will be key to solving the scientific questions that you find interesting.</p>
<p>At this stage, I’m sure you have many questions. Like, <em>What is “bioinformatics”?</em> <em>I’ve heard the phrase “Data Science”, is that what bioinformatics is?</em></p>
<dl class="footnote brackets">
<dt class="label" id="id7"><span class="brackets"><a class="fn-backref" href="#id1">3</a></span></dt>
<dd><p>I could write a very extensive essay on how open source software has, and continues to be, critical to science. Instead I have used the quote from Paul Romer to illuminate the fact that true scientific reproducibility requires transparency. By definition, open source software satisfies this condition while proprietary software does not.</p>
</dd>
<dt class="label" id="id8"><span class="brackets"><a class="fn-backref" href="#id2">4</a></span></dt>
<dd><p>Ok, Ok, I get it. This is a gross oversimplification of the differences between the languages.</p>
</dd>
</dl>
<p><span class="target" id="index-0"></span>Data Science is the joint application of algorithm development and statistical modelling to extracting information from what is referred to as “big data”. Bioinformatics, also referred to as computational biology, certainly fits within this rather loose definition. In essence, Bioinformatics is the union of algorithms and statistics focussed on extracting information from big biological data sets to advance knowledge of biological systems.</p>
<p><em>What will I get out of this course?</em> An understanding of how computer programs work. Improvements in your logical reasoning. An understanding of how to take advantage of computing to advance your own scientific interests. An appreciation of the pitfalls of software! All of this adds up to a superpower that will accelerate your work <a class="footnote-reference brackets" href="#id12" id="id11">5</a>.</p>
<dl class="footnote brackets">
<dt class="label" id="id12"><span class="brackets"><a class="fn-backref" href="#id11">5</a></span></dt>
<dd><p>Sorry, it’s bring your own cape.</p>
</dd>
</dl>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="acknowledgements.html" class="btn btn-neutral float-left" title="1. Acknowledgements" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="contribute.html" class="btn btn-neutral float-right" title="3. How to Contribute" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2022, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>