<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>8.1. Biological Data Formats &mdash; Topics in Bioinformatics 2022.08.25 documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="8.2. Alignments, Sequence Collections and Sequences" href="sequences.html" />
    <link rel="prev" title="8. Molecular Evolutionary Analyses Using cogent3" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">8.1. Biological Data Formats</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sequences-and-alignments">8.1.1. Sequences and alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#fasta">8.1.1.1. <code class="docutils literal notranslate"><span class="pre">fasta</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#id3">8.1.1.2. <code class="docutils literal notranslate"><span class="pre">genbank</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#gff-general-feature-format">8.1.1.3. <code class="docutils literal notranslate"><span class="pre">gff</span></code> – general feature format</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#newick-format-for-phylogenetic-trees">8.1.2. newick format for phylogenetic trees</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html">8.2. Alignments, Sequence Collections and Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#exercises">8.3. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="explore_variation.html">8.4. Exploring genetic variation using Alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="explore_variation.html#exercises">8.5. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="genetic_codes.html">8.6. Genetic Codes</a></li>
<li class="toctree-l2"><a class="reference internal" href="moltype.html">8.7. <code class="docutils literal notranslate"><span class="pre">MolType</span></code> – specifying biological sequence type</a></li>
<li class="toctree-l2"><a class="reference internal" href="apps.html">8.8. Overview of <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> apps</a></li>
<li class="toctree-l2"><a class="reference internal" href="using_models.html">8.9. Modelling sequence evolution</a></li>
<li class="toctree-l2"><a class="reference internal" href="using_models.html#exercises">8.10. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="tables.html">8.11. Tabular data</a></li>
<li class="toctree-l2"><a class="reference internal" href="statistical_tests.html">8.12. General purpose statistical tests using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">9. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">10. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">11. Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html"><span class="section-number">8. </span>Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a> &raquo;</li>
      <li><span class="section-number">8.1. </span>Biological Data Formats</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/cogent3/data_formats.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="biological-data-formats">
<h1><span class="section-number">8.1. </span>Biological Data Formats<a class="headerlink" href="#biological-data-formats" title="Permalink to this headline"></a></h1>
<figure style="padding: 4px;">
<img src="https://imgs.xkcd.com/comics/standards.png" style="height: auto; max-width=75%" alt="Editor Wars">
<figcaption>Read the original at <a href="https://xkcd.com/927/">XKCD</a></figcaption>
</figure><p>Oh dear, this is a mess. Not this document, but the “multitude” of data format standards in Bioinformatics. Bioinformatics is replete with different variants on basic formats. Not all standards are well documented and support for them is … well, I hope your experience is better than mine. The fact that data formats can be such a problem is a great argument for <a class="reference internal" href="../python/index.html#intro-to-python"><span class="std std-ref">knowing how to code</span></a>, since that skill gives you the ability to handle the different format variants that no existing tool handles.</p>
<p>The above aside, I will briefly describe and link <a class="footnote-reference brackets" href="#id2" id="id1">1</a> to the canonical description of a data format for the main types of data. I’ve grouped these into formats related to the major biological data types of sequences, genomic features, and phylogenetic trees.</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>If I can find one.</p>
</dd>
</dl>
<div class="section" id="sequences-and-alignments">
<h2><span class="section-number">8.1.1. </span>Sequences and alignments<a class="headerlink" href="#sequences-and-alignments" title="Permalink to this headline"></a></h2>
<div class="section" id="fasta">
<h3><span class="section-number">8.1.1.1. </span><code class="docutils literal notranslate"><span class="pre">fasta</span></code><a class="headerlink" href="#fasta" title="Permalink to this headline"></a></h3>
<p>The most commonly, and probably the easiest to parse, format is that of <code class="docutils literal notranslate"><span class="pre">fasta</span></code>. There are too many variants of this basic format, so I just <a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">link to the Wikipedia entry</a> and then present an example below.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seqlabel</span> <span class="mi">1</span> <span class="n">line</span>
<span class="n">ACCGGTGA</span>
<span class="n">AAG</span>
<span class="o">&gt;</span><span class="n">seqlabel2</span>
<span class="n">AGGCG</span>
</pre></div>
</div>
<p>While this is listed as a sequence format it’s often used to include multiple sequences, which may be aligned.</p>
</div>
<div class="section" id="id3">
<h3><span class="section-number">8.1.1.2. </span><code class="docutils literal notranslate"><span class="pre">genbank</span></code><a class="headerlink" href="#id3" title="Permalink to this headline"></a></h3>
<p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a> is a web portal to a multitude of valuable databases and bioinformatic tools. They also define a number of data file formats and, in particular, the <code class="docutils literal notranslate"><span class="pre">genbank</span></code> format for sequence data. This is format includes at a minimum extensive meta-data about a database entry. They do not always contain sequence, but will at least reference identifiers of related sequence records in their database. Below is a <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nucleotide/AF165912">sample record</a> from GenBank (which  I’ve edited for brevity)</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>LOCUS       AF165912                5485 bp    DNA     linear   PLN 29-JUL-1999
DEFINITION  Arabidopsis thaliana CTP:phosphocholine cytidylyltransferase (CCT)
            gene, complete cds.
ACCESSION   AF165912
VERSION     AF165912.1
KEYWORDS    .
SOURCE      Arabidopsis thaliana (thale cress)
  ORGANISM  Arabidopsis thaliana
            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;
            Spermatophyta; Magnoliophyta; eudicotyledons; Gunneridae;
            Pentapetalae; rosids; malvids; Brassicales; Brassicaceae;
            Camelineae; Arabidopsis.
REFERENCE   1  (bases 1 to 5485)
  AUTHORS   Choi,Y.H., Choi,S.B. and Cho,S.H.
  TITLE     Structure of a CTP:Phosphocholine Cytidylyltransferase Gene from
            Arabidopsis thaliana
  JOURNAL   Unpublished
REFERENCE   2  (bases 1 to 5485)
  AUTHORS   Choi,Y.H., Choi,S.B. and Cho,S.H.
  TITLE     Direct Submission
  JOURNAL   Submitted (06-JUL-1999) Biology, Inha University, Yonghyon-Dong
            253, Inchon 402-751, Korea
FEATURES             Location/Qualifiers
     source          1..5485
                     /organism=&quot;Arabidopsis thaliana&quot;
                     /mol_type=&quot;genomic DNA&quot;
                     /db_xref=&quot;taxon:3702&quot;
                     /ecotype=&quot;Col-0&quot;
     gene            1..4637
                     /gene=&quot;CCT&quot;
     regulatory      1..1602
                     /regulatory_class=&quot;promoter&quot;
                     /gene=&quot;CCT&quot;
&lt;features removed for brevity&gt;
ORIGIN
        1 ccagaatggt tactatggac atccgccaac catacaagct atggtgaaat gctttatcta
       61 tctcattttt agtttcaaag cttttgttat aacacatgca aatccatatc cgtaaccaat
      121 atccaatcgc ttgacatagt ctgatgaagt ttttggtagt taagataaag ctcgagactg
</pre></div>
</div>
</div>
<div class="section" id="gff-general-feature-format">
<h3><span class="section-number">8.1.1.3. </span><code class="docutils literal notranslate"><span class="pre">gff</span></code> – general feature format<a class="headerlink" href="#gff-general-feature-format" title="Permalink to this headline"></a></h3>
<blockquote class="epigraph">
<div><p>The GFF format, although widely used, has fragmented into multiple incompatible dialects.</p>
<p class="attribution">—Lincoln Stein, in the prologue to defining <code class="docutils literal notranslate"><span class="pre">gff3</span></code>.</p>
</div></blockquote>
<p>Sounds like the cartoon at the top doesn’t it! Anyway, it is widespread and an important format for storing information about sequence annotations <a class="footnote-reference brackets" href="#id5" id="id4">2</a>, so here’s the canonical <a class="reference external" href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">gff3</a> definition. This is a tab delimited format with 9 distinct fields. It’s the last field, <code class="docutils literal notranslate"><span class="pre">attributes</span></code>, that proves to be the most difficult to parse. Below is a small sample of a file posted on the definition page.</p>
<dl class="footnote brackets">
<dt class="label" id="id5"><span class="brackets"><a class="fn-backref" href="#id4">2</a></span></dt>
<dd><p>Annotations associate a reported feature of the sequence and the specific sequence coordinates to which it maps, e.g. the promoter for the gene CCT in the sample GenBank record.</p>
</dd>
</dl>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>0  ##gff-version 3.1.25
1  ##sequence-region ctg123 1 1497228
2  ctg123 . gene            1000  9000  .  +  .  ID=gene00001;Name=EDEN
3  ctg123 . TF_binding_site 1000  1012  .  +  .  ID=tfbs00001;Parent=gene00001
</pre></div>
</div>
</div>
</div>
<div class="section" id="newick-format-for-phylogenetic-trees">
<span id="newick"></span><h2><span class="section-number">8.1.2. </span>newick format for phylogenetic trees<a class="headerlink" href="#newick-format-for-phylogenetic-trees" title="Permalink to this headline"></a></h2>
<p>This is the most widespread text format for dsitributing phylogenetic trees. Clades of lineages are denoted by <code class="docutils literal notranslate"><span class="pre">()</span></code> and separate by <code class="docutils literal notranslate"><span class="pre">,</span></code> and can be grouped into subclades. Branch lengths are indicated by numbers after a colon character. There is some funky behaviour around dealing with spaces in tip names, they are often represented in the name as an underscore character (<code class="docutils literal notranslate"><span class="pre">&quot;_&quot;</span></code>). If you can, avoid any issues by not having spaces or underscores in names. Here’s a sample.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<p>((Human:0.006062440217780064,Chimpanzee:0.003020541234140796):0.09488527928524751,((Mouse:0.06659142318491332,Rat:0.05783486638653178):0.17244926332734278,Wombat:0.4522900123679113):0.0424545337445269,Horse:0.05802695948476483);</p>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="8. Molecular Evolutionary Analyses Using cogent3" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="sequences.html" class="btn btn-neutral float-right" title="8.2. Alignments, Sequence Collections and Sequences" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2022, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>