
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>8.7. Genetic Codes &#8212; Topics in Bioinformatics</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=fd3f3429" />
    <link rel="stylesheet" type="text/css" href="../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css?v=3ce756a4" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=75cd4315"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/thebelab-helper.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'cogent3/genetic_codes';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="8.8. Overview of cogent3 apps" href="apps.html" />
    <link rel="prev" title="8.5. Exploring genetic variation using Alignments" href="explore_variation.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Topics in Bioinformatics</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="data_formats.html">8.1. Biological Data Formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="moltype.html">8.2. <code class="docutils literal notranslate"><span class="pre">MolType</span></code> â€“ specifying biological sequence type</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html">8.3. Alignments, Sequence Collections and Sequences</a></li>

<li class="toctree-l2"><a class="reference internal" href="explore_variation.html">8.5. Exploring genetic variation using Alignments</a></li>

<li class="toctree-l2 current active"><a class="current reference internal" href="#">8.7. Genetic Codes</a></li>
<li class="toctree-l2"><a class="reference internal" href="apps.html">8.8. Overview of <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> apps</a></li>

<li class="toctree-l2"><a class="reference internal" href="using_models.html">8.10. Modelling sequence evolution</a></li>

<li class="toctree-l2"><a class="reference internal" href="tables.html">8.12. Tabular data</a></li>
<li class="toctree-l2"><a class="reference internal" href="statistical_tests.html">8.13. General purpose statistical tests using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../statistics/index.html">9. Statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/edit/main/cogent3/genetic_codes.rst" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/issues/new?title=Issue%20on%20page%20%2Fcogent3/genetic_codes.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/cogent3/genetic_codes.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Genetic Codes</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#useful-geneticcode-attributes">8.7.1. Useful <code class="docutils literal notranslate"><span class="pre">GeneticCode</span></code> attributes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#using-geneticcode-instances">8.7.2. Using <code class="docutils literal notranslate"><span class="pre">GeneticCode</span></code> instances</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercises">8.7.3. Exercises</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="genetic-codes">
<span id="id1"></span><h1><span class="section-number">8.7. </span>Genetic Codes<a class="headerlink" href="#genetic-codes" title="Link to this heading">#</a></h1>
<p>A genetic code is the fundamental information encoding system and a fascinating subject of study. <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> provides a dedicated object for handling genetic code information. The genetic codes included with <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> are indicated in the following table.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.25); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Specify a genetic code using either 'Name' or Code ID (as an integer or string)</span>
</caption>
<thead class="head_cell">
<th>Code ID</th><th>Name</th>
</thead>
<tbody>
<tr><td class="index"><span class="c3col_right">1</span></td><td><span class="c3col_left">Standard</span></td></tr>
<tr><td class="index"><span class="c3col_right">2</span></td><td><span class="c3col_left">Vertebrate Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">3</span></td><td><span class="c3col_left">Yeast Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">4</span></td><td><span class="c3col_left">Mold Mitochondrial; Protozoan Mitochondrial; Coelenterate Mitochondrial; Mycoplasma; Spiroplasma</span></td></tr>
<tr><td class="index"><span class="c3col_right">5</span></td><td><span class="c3col_left">Invertebrate Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">6</span></td><td><span class="c3col_left">Ciliate Nuclear; Dasycladacean Nuclear; Hexamita Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">9</span></td><td><span class="c3col_left">Echinoderm Mitochondrial; Flatworm Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">10</span></td><td><span class="c3col_left">Euplotid Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">11</span></td><td><span class="c3col_left">Bacterial, Archaeal and Plant Plastid</span></td></tr>
<tr><td class="index"><span class="c3col_right">12</span></td><td><span class="c3col_left">Alternative Yeast Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">13</span></td><td><span class="c3col_left">Ascidian Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">14</span></td><td><span class="c3col_left">Alternative Flatworm Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">15</span></td><td><span class="c3col_left">Blepharisma Macronuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">16</span></td><td><span class="c3col_left">Chlorophycean Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">21</span></td><td><span class="c3col_left">Trematode Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">22</span></td><td><span class="c3col_left">Scenedesmus obliquus Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">23</span></td><td><span class="c3col_left">Thraustochytrium Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">24</span></td><td><span class="c3col_left">Rhabdopleuridae Mitochondrial</span></td></tr>
<tr><td class="index"><span class="c3col_right">25</span></td><td><span class="c3col_left">Candidate Division SR1 and Gracilibacteria</span></td></tr>
<tr><td class="index"><span class="c3col_right">26</span></td><td><span class="c3col_left">Pachysolen tannophilus Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">27</span></td><td><span class="c3col_left">Karyorelict Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">28</span></td><td><span class="c3col_left">Condylostoma Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">29</span></td><td><span class="c3col_left">Mesodinium Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">30</span></td><td><span class="c3col_left">Peritrich Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">31</span></td><td><span class="c3col_left">Blastocrithidia Nuclear</span></td></tr>
<tr><td class="index"><span class="c3col_right">32</span></td><td><span class="c3col_left">Balanophoraceae Plastid</span></td></tr>
<tr><td class="index"><span class="c3col_right">33</span></td><td><span class="c3col_left">Cephalodiscidae Mitochondrial</span></td></tr>
</tbody>
</table>
<p>
27 rows x 2 columns</p>
</div></div></div>
</div>
<p>Use the top level <code class="docutils literal notranslate"><span class="pre">get_code()</span></code> function to get a specific genetic code. Use a code ID from the above table.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">get_code</span>

<span class="n">gc</span> <span class="o">=</span> <span class="n">get_code</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">gc</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Standard</span>
</caption>
<thead class="head_cell">
<th>aa</th><th>IUPAC code</th><th>codons</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Alanine</span></td><td><span class="c3col_left">A</span></td><td><span class="c3col_left">GCT,GCC,GCA,GCG</span></td></tr>
<tr><td><span class="c3col_left">Cysteine</span></td><td><span class="c3col_left">C</span></td><td><span class="c3col_left">TGT,TGC</span></td></tr>
<tr><td><span class="c3col_left">Aspartic Acid</span></td><td><span class="c3col_left">D</span></td><td><span class="c3col_left">GAT,GAC</span></td></tr>
<tr><td><span class="c3col_left">Glutamic Acid</span></td><td><span class="c3col_left">E</span></td><td><span class="c3col_left">GAA,GAG</span></td></tr>
<tr><td><span class="c3col_left">Phenylalanine</span></td><td><span class="c3col_left">F</span></td><td><span class="c3col_left">TTT,TTC</span></td></tr>
<tr><td><span class="c3col_left">Glycine</span></td><td><span class="c3col_left">G</span></td><td><span class="c3col_left">GGT,GGC,GGA,GGG</span></td></tr>
<tr><td><span class="c3col_left">Histidine</span></td><td><span class="c3col_left">H</span></td><td><span class="c3col_left">CAT,CAC</span></td></tr>
<tr><td><span class="c3col_left">Isoleucine</span></td><td><span class="c3col_left">I</span></td><td><span class="c3col_left">ATT,ATC,ATA</span></td></tr>
<tr><td><span class="c3col_left">Lysine</span></td><td><span class="c3col_left">K</span></td><td><span class="c3col_left">AAA,AAG</span></td></tr>
<tr><td><span class="c3col_left">Leucine</span></td><td><span class="c3col_left">L</span></td><td><span class="c3col_left">TTA,TTG,CTT,CTC,CTA,CTG</span></td></tr>
<tr><td><span class="c3col_left">Methionine</span></td><td><span class="c3col_left">M</span></td><td><span class="c3col_left">ATG</span></td></tr>
<tr><td><span class="c3col_left">Asparagine</span></td><td><span class="c3col_left">N</span></td><td><span class="c3col_left">AAT,AAC</span></td></tr>
<tr><td><span class="c3col_left">Proline</span></td><td><span class="c3col_left">P</span></td><td><span class="c3col_left">CCT,CCC,CCA,CCG</span></td></tr>
<tr><td><span class="c3col_left">Glutamine</span></td><td><span class="c3col_left">Q</span></td><td><span class="c3col_left">CAA,CAG</span></td></tr>
<tr><td><span class="c3col_left">Arginine</span></td><td><span class="c3col_left">R</span></td><td><span class="c3col_left">CGT,CGC,CGA,CGG,AGA,AGG</span></td></tr>
<tr><td><span class="c3col_left">Serine</span></td><td><span class="c3col_left">S</span></td><td><span class="c3col_left">TCT,TCC,TCA,TCG,AGT,AGC</span></td></tr>
<tr><td><span class="c3col_left">Threonine</span></td><td><span class="c3col_left">T</span></td><td><span class="c3col_left">ACT,ACC,ACA,ACG</span></td></tr>
<tr><td><span class="c3col_left">Valine</span></td><td><span class="c3col_left">V</span></td><td><span class="c3col_left">GTT,GTC,GTA,GTG</span></td></tr>
<tr><td><span class="c3col_left">Tryptophan</span></td><td><span class="c3col_left">W</span></td><td><span class="c3col_left">TGG</span></td></tr>
<tr><td><span class="c3col_left">Tyrosine</span></td><td><span class="c3col_left">Y</span></td><td><span class="c3col_left">TAT,TAC</span></td></tr>
<tr><td><span class="c3col_left">STOP</span></td><td><span class="c3col_left">*</span></td><td><span class="c3col_left">TAA,TAG,TGA</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<section id="useful-geneticcode-attributes">
<h2><span class="section-number">8.7.1. </span>Useful <code class="docutils literal notranslate"><span class="pre">GeneticCode</span></code> attributes<a class="headerlink" href="#useful-geneticcode-attributes" title="Link to this heading">#</a></h2>
<dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">sense_codons</span></code></dt><dd><p>The codons that encode an amino acid. The trinucleotide string is the key, and the single-character IUPAC code for the amino acid is the value.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">codons</span></code></dt><dd><p>Maps all codon strings to corresponding amino acid IUPAC code.</p>
</dd>
<dt><code class="docutils literal notranslate"><span class="pre">synonyms</span></code></dt><dd><p>Maps all amino acid IUPAC codes to their codons. (The reverse of <code class="docutils literal notranslate"><span class="pre">codons</span></code>.)</p>
</dd>
</dl>
</section>
<section id="using-geneticcode-instances">
<h2><span class="section-number">8.7.2. </span>Using <code class="docutils literal notranslate"><span class="pre">GeneticCode</span></code> instances<a class="headerlink" href="#using-geneticcode-instances" title="Link to this heading">#</a></h2>
<p>Genetic code objects act like dictionaries for trinucleotide strings or single-letter strings. The former is interpreted as RNA or DNA, the latter as the single-character amino acid code.</p>
<p>You can get the encoded amino acid from a RNA triplet</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aa</span> <span class="o">=</span> <span class="n">gc</span><span class="p">[</span><span class="s2">&quot;UAC&quot;</span><span class="p">]</span>
<span class="n">aa</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;Y&#39;
</pre></div>
</div>
</div>
</div>
<p>or DNA triplet.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aa</span> <span class="o">=</span> <span class="n">gc</span><span class="p">[</span><span class="s2">&quot;TAC&quot;</span><span class="p">]</span>
<span class="n">aa</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;Y&#39;
</pre></div>
</div>
</div>
</div>
<p>The mapping from codon to amino acid is provided by the genetic code instance <code class="docutils literal notranslate"><span class="pre">sense_codons</span></code> attribute. So, calling <code class="docutils literal notranslate"><span class="pre">list()</span></code> on that dict returns just the sense codons <a class="footnote-reference brackets" href="#id3" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id3" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">1</a><span class="fn-bracket">]</span></span>
<p>Iâ€™m truncating that list to just 4 elements to simplify the display.</p>
</aside>
</aside>
</aside>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">list</span><span class="p">(</span><span class="n">gc</span><span class="o">.</span><span class="n">sense_codons</span><span class="p">)[:</span><span class="mi">4</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;TTT&#39;, &#39;TTC&#39;, &#39;TTA&#39;, &#39;TTG&#39;]
</pre></div>
</div>
</div>
</div>
<p>You can get all the codons that encode an amino acid.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">codons</span> <span class="o">=</span> <span class="n">gc</span><span class="p">[</span><span class="s2">&quot;Y&quot;</span><span class="p">]</span>
<span class="n">codons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;TAT&#39;, &#39;TAC&#39;]
</pre></div>
</div>
</div>
</div>
<p>You can check whether a codon is a start</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gc</span><span class="o">.</span><span class="n">is_start</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>or stop codon</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gc</span><span class="o">.</span><span class="n">is_stop</span><span class="p">(</span><span class="s2">&quot;TAA&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>Stop codons are represented by <code class="docutils literal notranslate"><span class="pre">&quot;*&quot;</span></code> character.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gc</span><span class="p">[</span><span class="s2">&quot;TGA&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;*&#39;
</pre></div>
</div>
</div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gc</span><span class="p">[</span><span class="s2">&quot;*&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;TAA&#39;, &#39;TAG&#39;, &#39;TGA&#39;]
</pre></div>
</div>
</div>
</div>
<p>You can translate a string.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gc</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="s2">&quot;TCGACCGTTTAAGCC&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;STV*A&#39;
</pre></div>
</div>
</div>
</div>
<p>You can get the code as a Table,</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">table</span> <span class="o">=</span> <span class="n">gc</span><span class="o">.</span><span class="n">to_table</span><span class="p">()</span>
<span class="n">table</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Standard</span>
</caption>
<thead class="head_cell">
<th>aa</th><th>IUPAC code</th><th>codons</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Alanine</span></td><td><span class="c3col_left">A</span></td><td><span class="c3col_left">GCT,GCC,GCA,GCG</span></td></tr>
<tr><td><span class="c3col_left">Cysteine</span></td><td><span class="c3col_left">C</span></td><td><span class="c3col_left">TGT,TGC</span></td></tr>
<tr><td><span class="c3col_left">Aspartic Acid</span></td><td><span class="c3col_left">D</span></td><td><span class="c3col_left">GAT,GAC</span></td></tr>
<tr><td><span class="c3col_left">Glutamic Acid</span></td><td><span class="c3col_left">E</span></td><td><span class="c3col_left">GAA,GAG</span></td></tr>
<tr><td><span class="c3col_left">Phenylalanine</span></td><td><span class="c3col_left">F</span></td><td><span class="c3col_left">TTT,TTC</span></td></tr>
<tr><td><span class="c3col_left">Glycine</span></td><td><span class="c3col_left">G</span></td><td><span class="c3col_left">GGT,GGC,GGA,GGG</span></td></tr>
<tr><td><span class="c3col_left">Histidine</span></td><td><span class="c3col_left">H</span></td><td><span class="c3col_left">CAT,CAC</span></td></tr>
<tr><td><span class="c3col_left">Isoleucine</span></td><td><span class="c3col_left">I</span></td><td><span class="c3col_left">ATT,ATC,ATA</span></td></tr>
<tr><td><span class="c3col_left">Lysine</span></td><td><span class="c3col_left">K</span></td><td><span class="c3col_left">AAA,AAG</span></td></tr>
<tr><td><span class="c3col_left">Leucine</span></td><td><span class="c3col_left">L</span></td><td><span class="c3col_left">TTA,TTG,CTT,CTC,CTA,CTG</span></td></tr>
<tr><td><span class="c3col_left">Methionine</span></td><td><span class="c3col_left">M</span></td><td><span class="c3col_left">ATG</span></td></tr>
<tr><td><span class="c3col_left">Asparagine</span></td><td><span class="c3col_left">N</span></td><td><span class="c3col_left">AAT,AAC</span></td></tr>
<tr><td><span class="c3col_left">Proline</span></td><td><span class="c3col_left">P</span></td><td><span class="c3col_left">CCT,CCC,CCA,CCG</span></td></tr>
<tr><td><span class="c3col_left">Glutamine</span></td><td><span class="c3col_left">Q</span></td><td><span class="c3col_left">CAA,CAG</span></td></tr>
<tr><td><span class="c3col_left">Arginine</span></td><td><span class="c3col_left">R</span></td><td><span class="c3col_left">CGT,CGC,CGA,CGG,AGA,AGG</span></td></tr>
<tr><td><span class="c3col_left">Serine</span></td><td><span class="c3col_left">S</span></td><td><span class="c3col_left">TCT,TCC,TCA,TCG,AGT,AGC</span></td></tr>
<tr><td><span class="c3col_left">Threonine</span></td><td><span class="c3col_left">T</span></td><td><span class="c3col_left">ACT,ACC,ACA,ACG</span></td></tr>
<tr><td><span class="c3col_left">Valine</span></td><td><span class="c3col_left">V</span></td><td><span class="c3col_left">GTT,GTC,GTA,GTG</span></td></tr>
<tr><td><span class="c3col_left">Tryptophan</span></td><td><span class="c3col_left">W</span></td><td><span class="c3col_left">TGG</span></td></tr>
<tr><td><span class="c3col_left">Tyrosine</span></td><td><span class="c3col_left">Y</span></td><td><span class="c3col_left">TAT,TAC</span></td></tr>
<tr><td><span class="c3col_left">STOP</span></td><td><span class="c3col_left">*</span></td><td><span class="c3col_left">TAA,TAG,TGA</span></td></tr>
</tbody>
</table>
<p>
21 rows x 3 columns</p>
</div></div></div>
</div>
<p>See the <a class="reference external" href="https://cogent3.org">cogent3</a> cookbook documentation for more on using genetic codes.</p>
</section>
<section id="exercises">
<h2><span class="section-number">8.7.3. </span>Exercises<a class="headerlink" href="#exercises" title="Link to this heading">#</a></h2>
<p>Identify all sense codons that differ from each other at only one of the codon positions. Group these pairs by codon position <a class="footnote-reference brackets" href="#id5" id="id4" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. The following questions refer to these groupings.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id5" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">2</a><span class="fn-bracket">]</span></span>
<p>1st, 2nd and 3rd codon position.</p>
</aside>
</aside>
</aside>
<ol class="arabic">
<li><p>Pick a genetic code and, for each such codon position group, count the number of changes that are synonymous. Is there a difference between the codon position and the proportion of synonymous changes?</p></li>
<li><p>Does the property measured in the previous question differ between the genetic codes?</p></li>
<li><p>Categorise the codon differences by whether they are transition or transversion changes (see <a class="reference internal" href="../molevol/mutation.html#point-mutations"><span class="std std-ref">Point mutations</span></a>). Assess whether the fraction of synonymous changes differs between transition and transversion changes.</p></li>
<li><p>Is there variation (between the genetic codes) in the number of stop codons? Assess this programmatically.</p>
<blockquote>
<div><p><em>Hint: look at the attributes on the genetic code instance.</em></p>
</div></blockquote>
</li>
</ol>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="explore_variation.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">8.5. </span>Exploring genetic variation using Alignments</p>
      </div>
    </a>
    <a class="right-next"
       href="apps.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">8.8. </span>Overview of <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> apps</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#useful-geneticcode-attributes">8.7.1. Useful <code class="docutils literal notranslate"><span class="pre">GeneticCode</span></code> attributes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#using-geneticcode-instances">8.7.2. Using <code class="docutils literal notranslate"><span class="pre">GeneticCode</span></code> instances</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercises">8.7.3. Exercises</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gavin Huttley
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      Â© Copyright 2020-2024, Gavin Huttley.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>