<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>6.9. Measuring information using Shannon’s entropy &mdash; Topics in Bioinformatics 2022.09.14 documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="6.11. Position Specific Scoring Matrices (PSSMs)" href="pssm.html" />
    <link rel="prev" title="6.8. Finding motifs" href="motif_finding.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">6. Sequence Comparison</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="outline.html">6.1. Outline</a></li>
<li class="toctree-l2"><a class="reference internal" href="intro.html">6.2. Introduction to Comparison of Biological Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="encoding_information.html">6.3. The role of sequence in encoding life</a></li>
<li class="toctree-l2"><a class="reference internal" href="motifs.html">6.4. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="motifs.html#exercises">6.5. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_expression.html">6.6. Gene expression in a nutshell</a></li>
<li class="toctree-l2"><a class="reference internal" href="related_by_function.html">6.7. Experimental procedures for detecting functionally related sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="motif_finding.html">6.8. Finding motifs</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">6.9. Measuring information using Shannon’s entropy</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#measuring-entropy-for-a-simple-alignment">6.9.1. Measuring entropy for a simple alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#information-at-a-position">6.9.2. Information at a position</a></li>
<li class="toctree-l3"><a class="reference internal" href="#problems-with-sequence-logo">6.9.3. Problems with sequence logo</a></li>
<li class="toctree-l3"><a class="reference internal" href="#using-relative-entropy-as-an-alternate">6.9.4. Using relative entropy as an alternate</a></li>
<li class="toctree-l3"><a class="reference internal" href="#an-illustration-of-the-difference-identifying-neighbouring-base-effects-on-mutation">6.9.5. An illustration of the difference – identifying neighbouring base effects on mutation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#exercises">6.10. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html">6.11. Position Specific Scoring Matrices (PSSMs)</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html#exercises">6.12. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="odds_ratios.html">6.13. Odds-ratio’s</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html">6.14. Dotplot – Alignment of sequences related by descent from a common ancestor</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html#exercises">6.15. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html">6.16. Global Pairwise Alignment Using Dynamic Programming</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html#exercises">6.17. Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../statistics/index.html">9. Statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">12. Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html"><span class="section-number">6. </span>Sequence Comparison</a> &raquo;</li>
      <li><span class="section-number">6.9. </span>Measuring information using Shannon’s entropy</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/seqcomp/information.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="measuring-information-using-shannon-s-entropy">
<span id="shannon-entropy"></span><h1><span class="section-number">6.9. </span>Measuring information using Shannon’s entropy<a class="headerlink" href="#measuring-information-using-shannon-s-entropy" title="Permalink to this headline"></a></h1>
<p>We can measure the information content of “something” using a measurement called “entropy” (like a scale for information). The core unit is the “bit”, called a measure of surprise, which is the answer to a yes/no question.</p>
<p>Consider a state space containing only the letter <code class="docutils literal notranslate"><span class="pre">A</span></code>. We can never be “surprised” by the outcome of drawing a new letter from this space as it can only be of one type. So our entropy, in this case is 0. Add another letter and we have a chance to be “surprised”. So in general, as predictability goes up, entropy goes down.</p>
<p>As it turns out, entropy is a very fundamental quantity. You will see numerous expressions in statistics that relate to this. Given that the very purpose of genetical systems is to contain all the information sufficient to specify a living thing, application of statistical techniques built around the notion of “information” as a fundamental quantity seems quite natural.</p>
<p>There is a formal definition for entropy.</p>
<div class="math notranslate nohighlight">
\[H = -\sum_{i=0}^n p_i \log_2(p_i)\]</div>
<p>where <span class="math notranslate nohighlight">\(H\)</span> is Shannon entropy, <span class="math notranslate nohighlight">\(p\)</span> is a series of probabilities with <span class="math notranslate nohighlight">\(n\)</span> members and <span class="math notranslate nohighlight">\(p_i\)</span> the probability of state <span class="math notranslate nohighlight">\(i\)</span>. In words, entropy is calculated as the sum of <span class="math notranslate nohighlight">\(p_i\log_2(p_i)\)</span>. Further, in words the latter can be described as multiplying a probability by the log base 2 of itself.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><span class="math notranslate nohighlight">\(\sum_{i=0}^n\)</span> means take the sum of the series starting at index 0 and ending at index <span class="incremental">n</span>.</p>
</div>
<div class="section" id="measuring-entropy-for-a-simple-alignment">
<h2><span class="section-number">6.9.1. </span>Measuring entropy for a simple alignment<a class="headerlink" href="#measuring-entropy-for-a-simple-alignment" title="Permalink to this headline"></a></h2>
<p>Here’s the simple alignment from the previous section.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>TCAGA
TTCCA
TTCCA
TTTTC
TTTTC
</pre></div>
</div>
<p>Applying Shannon entropy to the sample alignment above gives:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Pos.   0     1     2     3     4
       T     C     A     G     A
       T     T     C     C     A
       T     T     C     C     A
       T     T     T     T     C
       T     T     T     T     C
Ent.   0    0.72  1.52  1.52  0.97
</pre></div>
</div>
<p>In the previous section, the display of coloured letters describing sequences that bind to TBP is referred to as a sequence logo <span id="id1">[<a class="reference internal" href="#id32" title="T D Schneider and R M Stephens. Sequence logos: a new way to display consensus sequences. Nucleic Acids Res, 18:6097-100, 1990. doi:10.1093/nar/18.20.6097.">SS90</a>]</span>. That representation is computed using position-wise entropy.</p>
</div>
<div class="section" id="information-at-a-position">
<span id="information"></span><h2><span class="section-number">6.9.2. </span>Information at a position<a class="headerlink" href="#information-at-a-position" title="Permalink to this headline"></a></h2>
<div class="math notranslate nohighlight">
\[R_i=\log_2(n) - H_i\]</div>
<p>where <span class="math notranslate nohighlight">\(R_i\)</span> is the information at position <em>i</em>, <em>n</em> is the number of possible states (4 for DNA), <span class="math notranslate nohighlight">\(\log_2(n)\)</span> is the maximum possible entropy, and <span class="math notranslate nohighlight">\(H_i\)</span> is the Shannon entropy for position <em>i</em>.</p>
<p>In this display, the letter height at a position is <span class="math notranslate nohighlight">\(p_{i,l} R_i\)</span>.</p>
</div>
<div class="section" id="problems-with-sequence-logo">
<h2><span class="section-number">6.9.3. </span>Problems with sequence logo<a class="headerlink" href="#problems-with-sequence-logo" title="Permalink to this headline"></a></h2>
<p>The background genomic distribution is assumed to be equally frequent nucleotides. Replacing <span class="math notranslate nohighlight">\(\log_2(n)\)</span> with the entropy of the genome can lead to <span class="math notranslate nohighlight">\(R&lt;0\)</span>.</p>
</div>
<div class="section" id="using-relative-entropy-as-an-alternate">
<h2><span class="section-number">6.9.4. </span>Using relative entropy as an alternate<a class="headerlink" href="#using-relative-entropy-as-an-alternate" title="Permalink to this headline"></a></h2>
<p>Relative entropy (<span class="incremental">RE</span>) is defined as</p>
<p><span class="math notranslate nohighlight">\(RE_i = \sum_{i=0}^n p_i \log_2(p_i / q_i)\)</span></p>
<p>where <span class="math notranslate nohighlight">\(q_i\)</span> is the background (e.g. genome) frequency of <em>i</em>.</p>
</div>
<div class="section" id="an-illustration-of-the-difference-identifying-neighbouring-base-effects-on-mutation">
<h2><span class="section-number">6.9.5. </span>An illustration of the difference – identifying neighbouring base effects on mutation<a class="headerlink" href="#an-illustration-of-the-difference-identifying-neighbouring-base-effects-on-mutation" title="Permalink to this headline"></a></h2>
<p><span class="math notranslate nohighlight">\(C\rightarrow T\)</span> mutations in vertebrates typically arise from deamination of 5-methyl-cytosine. The latter modification is typically applied to <span class="incremental">5’-CpG-3’</span>. Hence, there is a striking association of <span class="incremental">3’G</span> and this point mutation direction.</p>
<p>If we sample sequences that share this mutation in common, aligning them centred on the mutated position, we can assess which statistic better captures the neighbouring base association. What we <em>should</em> see is that the 3’ position should have high information which arises from the base <span class="incremental">G</span>.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><a class="reference internal image-reference" href="../_images/CtoT_MI.png"><img alt="../_images/CtoT_MI.png" src="../_images/CtoT_MI.png" style="width: 295.5px; height: 288.5px;" /></a>
</td>
<td><a class="reference internal image-reference" href="../_images/CtoT_RE.png"><img alt="../_images/CtoT_RE.png" src="../_images/CtoT_RE.png" style="width: 295.5px; height: 298.5px;" /></a>
</td>
</tr>
</tbody>
</table>
<p>These results originated from the honours research project of Ms Yicheng Zhu <span id="id4">[<a class="reference internal" href="#id31" title="Yicheng Zhu, Teresa Neeman, Von Bing Yap, and Gavin A Huttley. Statistical methods for identifying sequence motifs affecting point mutations. Genetics, 205:843-856, 2017. doi:10.1534/genetics.116.195677.">ZNYH17</a>]</span>.</p>
<div class="sidebar" id="math-log2">
<span id="index-0"></span><p class="sidebar-title">Math functions in python</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">numpy</span>

<span class="n">math</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">numpy</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>(1.0, 1.0)
</pre></div>
</div>
</div>
</div>
<p>In Python, mathematical functions are within the <code class="docutils literal notranslate"><span class="pre">math</span></code> module. These functions are typically also provided by the <code class="docutils literal notranslate"><span class="pre">numpy</span></code> library. The latter have the advantage of being able to operate on arrays.</p>
</div>
</div>
</div>
<div class="section" id="exercises">
<h1><span class="section-number">6.10. </span>Exercises<a class="headerlink" href="#exercises" title="Permalink to this headline"></a></h1>
<p>Defining a short sequence for use below.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;AGCCATTGCAAA&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<ol class="arabic">
<li><p>Write a function that converts <code class="docutils literal notranslate"><span class="pre">seq</span></code> into counts of nucleotides. (Use the standard library <code class="docutils literal notranslate"><span class="pre">collections.Counter</span></code> class.)</p></li>
<li><p>Write a function that takes a dict like object (e.g. the counter you’ve just created from (1)) of counts and converts it to numpy array whose length is 4.</p>
<blockquote>
<div><p>You should consider it critical that your function always returns arrays where the counts for each nucleotide are always in the same position. For instance, if you define the state order as T, C, A, G then the counts for T are always at index 0 – even if a T was not observed.</p>
<div class="tabbed-set docutils">
<input checked="checked" id="6c1b7734-d427-4f7a-9e0a-960507548318" name="79d6a3a4-156b-425e-8e0f-4f47ccc1182c" type="radio">
</input><label class="tabbed-label" for="6c1b7734-d427-4f7a-9e0a-960507548318">
Expected output</label><div class="tabbed-content docutils">
<p>If you select that nucleotide ordering, your function should return the nfollowing given the nucleotide counts <code class="docutils literal notranslate"><span class="pre">dict</span></code> from <code class="docutils literal notranslate"><span class="pre">seq</span></code> as input</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>array([2, 3, 5, 2])
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div></blockquote>
</li>
<li><p>Write a function that converts these counts into a frequencies.</p></li>
<li><p>Converting the mathematical expression for entropy into a Python function. Apply it to the results from applying the functions you wrote for (1-3) to <code class="docutils literal notranslate"><span class="pre">seq</span></code> (<a class="reference internal" href="#math-log2"><span class="std std-ref">for taking log2 see the math functions sidebar</span></a>).</p></li>
<li><p>Apply the functions you’ve written to the following randomly generated sequence.</p>
<blockquote>
<div><div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;GGAGACGCCCGCGACGGAGGAAAACAAGCCGAGGGGACGGGGGCGAGCAG&#39;
</pre></div>
</div>
</div>
</div>
<p>What do you observe? Why is it happening?</p>
</div></blockquote>
</li>
<li><p>Modify your function for answer (3) so, if there are zero values, it adds a pseudocount to <strong>all</strong> of the array elements before converting to frequencies. Set a default pseudocount of 0.5.</p>
<blockquote>
<div><p>Make sure your frequencies sum to 1.</p>
<p>Is there another approach to solving this issue?</p>
</div></blockquote>
</li>
</ol>
<hr class="docutils" />
<p class="rubric">Citations</p>
<div class="docutils container" id="id5">
<dl class="citation">
<dt class="label" id="id32"><span class="brackets"><a class="fn-backref" href="#id1">SS90</a></span></dt>
<dd><p>T D Schneider and R M Stephens. Sequence logos: a new way to display consensus sequences. <em>Nucleic Acids Res</em>, 18:6097–100, 1990. <a class="reference external" href="https://doi.org/10.1093/nar/18.20.6097">doi:10.1093/nar/18.20.6097</a>.</p>
</dd>
<dt class="label" id="id31"><span class="brackets"><a class="fn-backref" href="#id4">ZNYH17</a></span></dt>
<dd><p>Yicheng Zhu, Teresa Neeman, Von Bing Yap, and Gavin A Huttley. Statistical methods for identifying sequence motifs affecting point mutations. <em>Genetics</em>, 205:843–856, 2017. <a class="reference external" href="https://doi.org/10.1534/genetics.116.195677">doi:10.1534/genetics.116.195677</a>.</p>
</dd>
</dl>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="motif_finding.html" class="btn btn-neutral float-left" title="6.8. Finding motifs" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pssm.html" class="btn btn-neutral float-right" title="6.11. Position Specific Scoring Matrices (PSSMs)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2022, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>