<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>6.2. Introduction to Comparison of Biological Sequences &mdash; Topics in Bioinformatics 2022.07.28 documentation</title><link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
      <link rel="stylesheet" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="6.3. The role of sequence in encoding life" href="encoding_information.html" />
    <link rel="prev" title="6.1. Outline" href="outline.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">6. Sequence Comparison</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="outline.html">6.1. Outline</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">6.2. Introduction to Comparison of Biological Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#descended-from-a-common-ancestor">6.2.1. Descended from a common ancestor</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#the-major-take-home-messages-relevant-to-this-topic">6.2.1.1. The major take-home messages relevant to this topic</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="encoding_information.html">6.3. The role of sequence in encoding life</a></li>
<li class="toctree-l2"><a class="reference internal" href="motifs.html">6.4. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="motifs.html#exercises">6.5. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_expression.html">6.6. Gene expression in a nutshell</a></li>
<li class="toctree-l2"><a class="reference internal" href="related_by_function.html">6.7. Experimental procedures for detecting functionally related sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="motif_finding.html">6.8. Finding motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html">6.9. Measuring information using Shannon’s entropy</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html#exercises">6.10. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html">6.11. Position Specific Scoring Matrices (PSSMs)</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html#exercises">6.12. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="odds_ratios.html">6.13. Odds-ratio’s</a></li>
<li class="toctree-l2"><a class="reference internal" href="statistics.html">6.14. Refresher on statistical inference</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html">6.15. Dotplot – Alignment of sequences related by descent from a common ancestor</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html#exercises">6.16. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html">6.17. Global Pairwise Alignment Using Dynamic Programming</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html#exercises">6.18. Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">9. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">10. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">11. Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html"><span class="section-number">6. </span>Sequence Comparison</a> &raquo;</li>
      <li><span class="section-number">6.2. </span>Introduction to Comparison of Biological Sequences</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/seqcomp/intro.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="introduction-to-comparison-of-biological-sequences">
<h1><span class="section-number">6.2. </span>Introduction to Comparison of Biological Sequences<a class="headerlink" href="#introduction-to-comparison-of-biological-sequences" title="Permalink to this headline"></a></h1>
<p>The comparison of biological sequences is arguably the core task of bioinformatics. In this course, we split this into two domains: comparison of sequences that are related solely by function; comparison of sequences that share a common ancestor.</p>
<p>We start the topic by framing the problem of sequence analysis from the perspective of what we understand about the origins of biological sequences and the role they play in living systems. Our purpose with these descriptions is not to provide an exhaustive background but, hopefully, just a refresher of things you already know.</p>
<div class="section" id="descended-from-a-common-ancestor">
<h2><span class="section-number">6.2.1. </span>Descended from a common ancestor<a class="headerlink" href="#descended-from-a-common-ancestor" title="Permalink to this headline"></a></h2>
<p>In 1859, Charles Darwin (aka the Dude!) published his seminal work on the origin of species. In it, he proposed that descent occurred with modification and that this “modification” was heritable variation which affected an individuals ability to survive and reproduce. Evolution thus occurring as a consequence of natural selection eliminating individuals that contained variants not well suited to their environment and / or had a lower ability to reproduce when compared to their contemporaries.</p>
<p>Darwin clearly understood that organismal phenotype was heritable and thus required an information system with associated mechanisms for its transmission. He further understood that differences existed between individuals in the copy of the information system they possessed which must be responsible for those differences in phenotype. But he did not understand the mechanism of how inheritance occurred <a class="footnote-reference brackets" href="#id2" id="id1">1</a>.</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Well he did have a model, but it was disproved by his cousin, Francis Galton, a major figure in the history of statistics.</p>
</dd>
</dl>
<p>Darwin’s work was highly controversial at the time, including his conclusion that all living things on Earth had descended from a common ancestor.</p>
<p>Just 6 years later, in 1865, Gregor Mendel published his work on the laws of inheritance. These constitute the foundation principles of what we now refer to as genetics <a class="footnote-reference brackets" href="#id4" id="id3">2</a>. Unfortunately, Mendel’s work was largely ignored for 35 years until after it was replicated by 3 botanists (de Vries, Correns, and von Tschermak-Seysenegg). It took another 53 years before Watson and Crick presented evidence for the physical structure of the hereditary molecule – DNA. This laid the foundation for examination of how information for life is encoded.</p>
<dl class="footnote brackets">
<dt class="label" id="id4"><span class="brackets"><a class="fn-backref" href="#id3">2</a></span></dt>
<dd><p>Mendel did not use that term.</p>
</dd>
</dl>
<p>In the 1920-40’s, seminal work done by <a class="reference external" href="https://en.wikipedia.org/wiki/Sewall_Wright">Sewall Wright</a>, <a class="reference external" href="https://en.wikipedia.org/wiki/J._B._S._Haldane">J. B. S. Haldane</a>,  <a class="reference external" href="https://en.wikipedia.org/wiki/Ronald_Fisher">Ronald Fisher</a>, <a class="reference external" href="https://en.wikipedia.org/wiki/Theodosius_Dobzhansky">Theodosius Dobzhansky</a> and others established what is referred to as the “Neo-Darwinian Synthesis”. This work re-expressed Darwin’s theory in terms of genetical mechanisms of inheritance – unifying the two fields. The work is dominantly mathematical, it constitutes the foundations of population genetics, and it provides the basis from which we now view the distribution of genetic variation. Another major extension to those foundations from which we interpret genetic variation was the work of Motoo Kimura in 1968 on the Neutral Theory of Molecular Evolution. (As in all the above, there were numerous other contributors to development of these ideas and theory.)</p>
<p>These large frameworks provide a perspective from within which we can view biological systems, irrespective of the scale we are interested in. Ranging from biochemical reactions, to cellular biology to organisms to ecosystems. All levels of biological organisation have been shaped by the processes that govern the inheritance of the molecular information system and by the relationship between the information encoded in DNA and its manifestation in phenotype.</p>
<p>In this topic we will begin thinking about how we can interrogate DNA sequences to discover how information is encoded and how we can quantify relationships among DNA sequences. The latter is a precursor to a mechanistic understanding of how evolution has, and is, operating.</p>
<div class="section" id="the-major-take-home-messages-relevant-to-this-topic">
<h3><span class="section-number">6.2.1.1. </span>The major take-home messages relevant to this topic<a class="headerlink" href="#the-major-take-home-messages-relevant-to-this-topic" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>All living things on the planet ultimately descended from a common ancestor</p></li>
<li><p>Both parents transmit hereditary information to progeny <a class="footnote-reference brackets" href="#id6" id="id5">3</a></p></li>
<li><p>Hereditary information is encoded in a nucleic acid, typically DNA</p></li>
<li><p>Each position in a DNA sequence derived from a residue in an ancestor</p></li>
</ul>
<dl class="footnote brackets">
<dt class="label" id="id6"><span class="brackets"><a class="fn-backref" href="#id5">3</a></span></dt>
<dd><p>For diploid, sexually reproducing organisms.</p>
</dd>
</dl>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="outline.html" class="btn btn-neutral float-left" title="6.1. Outline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="encoding_information.html" class="btn btn-neutral float-right" title="6.3. The role of sequence in encoding life" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2022, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>