
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>6.6. \(k\)-mers and Motifs &#8212; Topics in Bioinformatics</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=fd3f3429" />
    <link rel="stylesheet" type="text/css" href="../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css?v=3ce756a4" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=75cd4315"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/thebelab-helper.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'seqcomp/motifs';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="6.8. Experimental procedures for detecting functionally related sequences" href="related_by_function.html" />
    <link rel="prev" title="6.4. Gene expression in a nutshell" href="gene_expression.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Topics in Bioinformatics</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">6. Sequence Comparison</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="outline.html">6.1. Outline</a></li>
<li class="toctree-l2"><a class="reference internal" href="intro.html">6.2. Introduction to Comparison of Biological Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="encoding_information.html">6.3. The role of sequence in encoding life</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_expression.html">6.4. Gene expression in a nutshell</a></li>

<li class="toctree-l2 current active"><a class="current reference internal" href="#">6.6. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a></li>

<li class="toctree-l2"><a class="reference internal" href="related_by_function.html">6.8. Experimental procedures for detecting functionally related sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="motif_finding.html">6.9. Finding motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html">6.10. Measuring information using Shannon’s entropy</a></li>

<li class="toctree-l2"><a class="reference internal" href="pssm.html">6.12. Position Specific Scoring Matrices (PSSMs)</a></li>

<li class="toctree-l2"><a class="reference internal" href="odds_ratios.html">6.14. Odds-ratio’s</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html">6.15. Dotplot – Alignment of sequences related by descent from a common ancestor</a></li>

<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html">6.17. Global Pairwise Alignment Using Dynamic Programming</a></li>

</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../statistics/index.html">9. Statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/edit/main/seqcomp/motifs.rst" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/issues/new?title=Issue%20on%20page%20%2Fseqcomp/motifs.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/seqcomp/motifs.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>k-mers and Motifs</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">6.6. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#kmers">6.6.1. <span class="math notranslate nohighlight">\(k\)</span>-mers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motifs">6.6.2. Motifs</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#exercises">6.7. Exercises</a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="kmers-and-motifs">
<h1><span class="section-number">6.6. </span><span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs<a class="headerlink" href="#kmers-and-motifs" title="Link to this heading">#</a></h1>
<section id="kmers">
<span id="index-0"></span><h2><span class="section-number">6.6.1. </span><span class="math notranslate nohighlight">\(k\)</span>-mers<a class="headerlink" href="#kmers" title="Link to this heading">#</a></h2>
<p>One way of describing DNA sequences is simply counting the occurrences of DNA words of a fixed size. Such words are often referred to as <span class="math notranslate nohighlight">\(k\)</span>-mers where <span class="math notranslate nohighlight">\(k\)</span> refers to the word length. If we set <span class="math notranslate nohighlight">\(k=1\)</span>, then the complete set of possible 1-mers is the listing of the nucleotides, i.e. <code class="docutils literal notranslate"><span class="pre">{A,</span> <span class="pre">C,</span> <span class="pre">G,</span> <span class="pre">T}</span></code>.</p>
<p>But why do it? Why is knowing <span class="math notranslate nohighlight">\(k\)</span>-mers informative? There are a multitude of ways in which <span class="math notranslate nohighlight">\(k\)</span>-mer counting is used in bioinformatics. One concerns evolutionary relatedness, and another concerns functional encoding. Before we tackle those, let’s go ahead and define <span class="math notranslate nohighlight">\(k\)</span>-mers in more detail.</p>
<p>We define the set of possible <span class="math notranslate nohighlight">\(k\)</span>-mers as the full enumeration of characters in a state space. In English, since DNA has four letters there are <span class="math notranslate nohighlight">\(4^k\)</span> distinct <span class="math notranslate nohighlight">\(k\)</span>-mers. So for <span class="math notranslate nohighlight">\(k=2\)</span>, that’s <span class="math notranslate nohighlight">\(4^2=16\)</span>. These are all the possible dinucleotides.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AA, AC, AG, AT, CA, CC, CG, CT, GA, GC, GG, GT, TA, TC, TG, TT
</pre></div>
</div>
</div>
</div>
<p>Setting <span class="math notranslate nohighlight">\(k=3\)</span>, we have 64 possible trinucleotides, for <span class="math notranslate nohighlight">\(k=4\)</span>, we have 256 tetranucleotides and so on.</p>
<p>We typically define the <span class="math notranslate nohighlight">\(k\)</span>-mer distribution of a biological sequence as the counts of all possible occurrences of each <span class="math notranslate nohighlight">\(k\)</span>-mer. Consider the following 11bp long DNA sequence.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;ACGGTGCCAGG&#39;
</pre></div>
</div>
</div>
</div>
<p>The distribution of <span class="math notranslate nohighlight">\(k\)</span>-mers for <span class="math notranslate nohighlight">\(k=1\)</span> is</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;A&#39;: 2, &#39;C&#39;: 3, &#39;G&#39;: 5, &#39;T&#39;: 1}
</pre></div>
</div>
</div>
</div>
<p>Each value in the <code class="docutils literal notranslate"><span class="pre">dict</span></code> corresponds to the number of occurrences of that nucleotide.</p>
<p>The distribution of <span class="math notranslate nohighlight">\(k\)</span>-mers for <span class="math notranslate nohighlight">\(k=2\)</span> is</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;AC&#39;: 1, &#39;CG&#39;: 1, &#39;GG&#39;: 2, &#39;GT&#39;: 1, &#39;TG&#39;: 1, &#39;GC&#39;: 1, &#39;CC&#39;: 1, &#39;CA&#39;: 1, &#39;AG&#39;: 1}
</pre></div>
</div>
</div>
</div>
<p>where each value corresponds to the number of occurrences of that dinucleotide. If you sum the values you get 10, from a sequence of length 11. Wait, what … how does that work! Surely there are only 5 dinucleotides in a sequence of length 11!</p>
<p>We derive a <span class="math notranslate nohighlight">\(k\)</span>-mer distribution by considering <strong>all possible words</strong> of length <span class="math notranslate nohighlight">\(k\)</span>, Let’s simplify the problem by working on just the first 3 nucleotides, <code class="docutils literal notranslate"><span class="pre">ACG</span></code>. This has two possible dinucleotides – <code class="docutils literal notranslate"><span class="pre">AC</span></code> and <code class="docutils literal notranslate"><span class="pre">CG</span></code>. Thus, we slide the start position along the sequence by 1 and take the <span class="math notranslate nohighlight">\(k\)</span> nucleotides from that position (<a class="reference internal" href="#id1"><span class="std std-ref">see Sliding window</span></a>).</p>
<aside class="margin sidebar" id="id1">
<p class="sidebar-title">Sliding window</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Index</span>      <span class="mi">012</span>
<span class="n">Sequence</span>   <span class="n">ACG</span>
<span class="n">Index</span> <span class="mi">0</span>    <span class="n">AC</span>
<span class="n">Index</span> <span class="mi">1</span>     <span class="n">CG</span>
</pre></div>
</div>
<p>To identify all dinucleotides (<span class="math notranslate nohighlight">\(k=2\)</span>), we iterate over all indices, slice the sequence from the current index to be length 2. Clearly, we need to make sure to get the last <span class="math notranslate nohighlight">\(k\)</span>-mer correct, i.e. it must also be of length 2.</p>
</aside>
<p>So how do <span class="math notranslate nohighlight">\(k\)</span>-mers get used? They are used as a measure of evolutionary relatedness. The “distance” between <span class="math notranslate nohighlight">\(k\)</span>-mer distributions of closely related species is typically smaller than the distance between distantly related species. The genomic signature statistics of <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/9294192/">Karlin and colleagues</a> are related to these measures. Moreover, <span class="math notranslate nohighlight">\(k\)</span>-mers are <a class="reference external" href="https://www.genetics.org/content/215/1/25">employed in machine learning algorithms</a> such as that of Zhu et al (from my own lab).</p>
</section>
<section id="motifs">
<span id="index-1"></span><h2><span class="section-number">6.6.2. </span>Motifs<a class="headerlink" href="#motifs" title="Link to this heading">#</a></h2>
<p>Another application of <span class="math notranslate nohighlight">\(k\)</span>-mers is in how they relate to the concept of motifs. A motif is a short sequence that occurs multiple times in a DNA, RNA or protein sequence. The phrase can also be applied quite generally. For instance, the <code class="docutils literal notranslate"><span class="pre">&quot;ATG&quot;</span></code> motif corresponds to the start codon of protein coding genes. The phrase is typically applied to sequences that have some functional association. Arguably, this notion is exemplified by sequence logos, a statistical method used for visualisation of motifs. The DNA binding motif of the TBP protein is visually represented by a <a class="reference internal" href="gene_expression.html#binding-to-dna"><span class="std std-ref">sequence logo</span></a>. A related visualisation technique was developed by Zhu et al (from my own lab) for identifying <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/27974498">sequence motifs associated with mutation processes</a>.</p>
<p>As the start codon example illustrates, motifs represent a fundamental concept in the description of information encoding by DNA sequences.</p>
</section>
</section>
<section id="exercises">
<h1><span class="section-number">6.7. </span>Exercises<a class="headerlink" href="#exercises" title="Link to this heading">#</a></h1>
<ol class="arabic">
<li><p>Consider the sequence <code class="docutils literal notranslate"><span class="pre">seq</span></code>. How many <span class="math notranslate nohighlight">\(k\)</span>-mers are there for <span class="math notranslate nohighlight">\(k=1,2,3\)</span>?</p>
<blockquote>
<div><div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;ACG&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div></blockquote>
</li>
<li><p>For a sequence of length 7, how many <span class="math notranslate nohighlight">\(k\)</span>-mers are there for <span class="math notranslate nohighlight">\(k=1,2,3\)</span>?</p></li>
<li><p>Write an equation for the number of <span class="math notranslate nohighlight">\(k\)</span>-mers in a sequence of length <span class="math notranslate nohighlight">\(n\)</span>. When you set <span class="math notranslate nohighlight">\(n=3, 7\)</span> and <span class="math notranslate nohighlight">\(k=1, 2, 3\)</span> you should get the same answers as above.</p></li>
<li><p>Write an algorithm that produces all the dinucleotides for <code class="docutils literal notranslate"><span class="pre">seq</span></code>.</p></li>
<li><p>Then do it for all the trinucleotides in <code class="docutils literal notranslate"><span class="pre">seq</span></code>.</p></li>
<li><p>The Python standard library has lots of very useful goodies. Investigate the <code class="docutils literal notranslate"><span class="pre">Counter</span></code> class from <code class="docutils literal notranslate"><span class="pre">collections</span></code> and apply it to your result from (4) and (5) to determine the <span class="math notranslate nohighlight">\(k\)</span>-mer counts. (Use google!)</p></li>
</ol>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="gene_expression.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">6.4. </span>Gene expression in a nutshell</p>
      </div>
    </a>
    <a class="right-next"
       href="related_by_function.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">6.8. </span>Experimental procedures for detecting functionally related sequences</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">6.6. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#kmers">6.6.1. <span class="math notranslate nohighlight">\(k\)</span>-mers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#motifs">6.6.2. Motifs</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#exercises">6.7. Exercises</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gavin Huttley
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2020-2024, Gavin Huttley.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>