

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>6.4. \(k\)-mers and Motifs &mdash; Topics in Bioinformatics 2021.08.24 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
  <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
  <link rel="stylesheet" href="../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
  <link rel="stylesheet" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
  <link rel="stylesheet" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="6.6. Gene expression in a nutshell" href="gene_expression.html" />
    <link rel="prev" title="6.3. The role of sequence in encoding life" href="encoding_information.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> Topics in Bioinformatics
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">6. Sequence Comparison</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="outline.html">6.1. Outline</a></li>
<li class="toctree-l2"><a class="reference internal" href="intro.html">6.2. Introduction to Comparison of Biological Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="encoding_information.html">6.3. The role of sequence in encoding life</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">6.4. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#kmers">6.4.1. <span class="math notranslate nohighlight">\(k\)</span>-mers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#motifs">6.4.2. Motifs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#exercises">6.5. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_expression.html">6.6. Gene expression in a nutshell</a></li>
<li class="toctree-l2"><a class="reference internal" href="related_by_function.html">6.7. Experimental procedures for detecting functionally related sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="motif_finding.html">6.8. Finding motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html">6.9. Measuring information using Shannon’s entropy</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html#exercises">6.10. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html">6.11. Position Specific Scoring Matrices (PSSMs)</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html#exercises">6.12. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="odds_ratios.html">6.13. Odds-ratio’s</a></li>
<li class="toctree-l2"><a class="reference internal" href="statistics.html">6.14. Refresher on statistical inference</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html">6.15. Dotplot – Alignment of sequences related by descent from a common ancestor</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html#exercises">6.16. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html">6.17. Global Pairwise Alignment Using Dynamic Programming</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html#exercises">6.18. Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">9. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">10. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">11. Index</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html"><span class="section-number">6. </span>Sequence Comparison</a> &raquo;</li>
        
      <li><span class="section-number">6.4. </span><span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/seqcomp/motifs.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="kmers-and-motifs">
<h1><span class="section-number">6.4. </span><span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs<a class="headerlink" href="#kmers-and-motifs" title="Permalink to this headline">¶</a></h1>
<div class="section" id="kmers">
<span id="index-0"></span><h2><span class="section-number">6.4.1. </span><span class="math notranslate nohighlight">\(k\)</span>-mers<a class="headerlink" href="#kmers" title="Permalink to this headline">¶</a></h2>
<p>One way of describing DNA sequences is simply counting the occurrences of DNA words of a fixed size. Such words are often referred to as <span class="math notranslate nohighlight">\(k\)</span>-mers where <span class="math notranslate nohighlight">\(k\)</span> refers to the word length. If we set <span class="math notranslate nohighlight">\(k=1\)</span>, then the complete set of possible 1-mers is the listing of the nucleotides, i.e. <code class="docutils literal notranslate"><span class="pre">{A,</span> <span class="pre">C,</span> <span class="pre">G,</span> <span class="pre">T}</span></code>.</p>
<p>But why do it? Why is knowing <span class="math notranslate nohighlight">\(k\)</span>-mers informative? There are a multitude of ways in which <span class="math notranslate nohighlight">\(k\)</span>-mer counting is used in bioinformatics. One concerns evolutionary relatedness, another concerns functional encoding. Before we tackle those, lets go ahead and define <span class="math notranslate nohighlight">\(k\)</span>-mers in more detail.</p>
<p>We define the set of possible <span class="math notranslate nohighlight">\(k\)</span>-mers as the full enumeration of a characters in a state space. In English, since DNA has four letters there are <span class="math notranslate nohighlight">\(4^k\)</span> distinct <span class="math notranslate nohighlight">\(k\)</span>-mers. So for <span class="math notranslate nohighlight">\(k=2\)</span>, that’s <span class="math notranslate nohighlight">\(4^2=16\)</span>. These are all the possible dinucleotides.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AA, AC, AG, AT, CA, CC, CG, CT, GA, GC, GG, GT, TA, TC, TG, TT
</pre></div>
</div>
</div>
</div>
<p>Setting <span class="math notranslate nohighlight">\(k=3\)</span>, we have 64 possible trinucleotides, for <span class="math notranslate nohighlight">\(k=4\)</span>, we have 256 tetranucleotides and so on.</p>
<p>We typically define the <span class="math notranslate nohighlight">\(k\)</span>-mer distribution of a biological sequence as the counts all possible occurrences of each <span class="math notranslate nohighlight">\(k\)</span>-mer. Consider the following 11bp long DNA sequence.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>&#39;ACGGTGCCAGG&#39;
</pre></div>
</div>
</div>
</div>
<p>The distribution of <span class="math notranslate nohighlight">\(k\)</span>-mers for <span class="math notranslate nohighlight">\(k=1\)</span> is</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;A&#39;: 2, &#39;C&#39;: 3, &#39;G&#39;: 5, &#39;T&#39;: 1}
</pre></div>
</div>
</div>
</div>
<p>Each value in the <code class="docutils literal notranslate"><span class="pre">dict</span></code> corresponds to the number of occurrences of that nucleotide.</p>
<p>The distribution of <span class="math notranslate nohighlight">\(k\)</span>-mers for <span class="math notranslate nohighlight">\(k=2\)</span> is</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;AC&#39;: 1, &#39;CG&#39;: 1, &#39;GG&#39;: 2, &#39;GT&#39;: 1, &#39;TG&#39;: 1, &#39;GC&#39;: 1, &#39;CC&#39;: 1, &#39;CA&#39;: 1, &#39;AG&#39;: 1}
</pre></div>
</div>
</div>
</div>
<p>where each value corresponds to the number of occurrences of that dinucleotide. If you sum the values you get 10, from a sequence of length 11. Wait, what … how does that work! Surely there are only 5 dinucleotides in a sequence of length 11!</p>
<p>We derive a <span class="math notranslate nohighlight">\(k\)</span>-mer distribution by considering <strong>all possible words</strong> of length <span class="math notranslate nohighlight">\(k\)</span>, Let’s simplify the problem by working on just the first 3 nucleotides, <code class="docutils literal notranslate"><span class="pre">ACG</span></code>. This has two possible dinucleotides – <code class="docutils literal notranslate"><span class="pre">AC</span></code> and <code class="docutils literal notranslate"><span class="pre">CG</span></code>. Thus, we slide the start position along the sequence by 1 and take the <span class="math notranslate nohighlight">\(k\)</span> nucleotides from that position (<a class="reference internal" href="#id1"><span class="std std-ref">see Sliding window</span></a>).</p>
<div class="sidebar" id="id1">
<p class="sidebar-title">Sliding window</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Index</span>      <span class="mi">012</span>
<span class="n">Sequence</span>   <span class="n">ACG</span>
<span class="n">Index</span> <span class="mi">0</span>    <span class="n">AC</span>
<span class="n">Index</span> <span class="mi">1</span>     <span class="n">CG</span>
</pre></div>
</div>
<p>To identify all dinucleotides (<span class="math notranslate nohighlight">\(k=2\)</span>), we iterate over all indices, slice the sequence from the current index to be length 2. Clearly, we need to make sure to get the last <span class="math notranslate nohighlight">\(k\)</span>-mer correct, i.e. it must also be of length 2.</p>
</div>
<p>So how do <span class="math notranslate nohighlight">\(k\)</span>-mers get used? They are used as a measure of evolutionary relatedness. The “distance” between <span class="math notranslate nohighlight">\(k\)</span>-mer distributions of closely related species is typically smaller than the distance between distantly related species. The genomic signature statistics of <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/9294192/">Karlin and colleagues</a> are related to these measures. Moreover, <span class="math notranslate nohighlight">\(k\)</span>-mers are <a class="reference external" href="https://www.genetics.org/content/215/1/25">employed in machine learning algorithms</a> such as that of Zhu et al (from my own lab).</p>
</div>
<div class="section" id="motifs">
<span id="index-1"></span><h2><span class="section-number">6.4.2. </span>Motifs<a class="headerlink" href="#motifs" title="Permalink to this headline">¶</a></h2>
<p>Another application of <span class="math notranslate nohighlight">\(k\)</span>-mers is in how they relate to the concept of motifs. A motif is a short sequence that occurs multiple times in a DNA, RNA or protein sequence. The phrase can also be applied quite generally. For instance, the <code class="docutils literal notranslate"><span class="pre">&quot;ATG&quot;</span></code> motif corresponds to the start codon of protein coding genes. The phrase is typically applied to sequences that have some functional association. Arguably, this notion is exemplified by sequence logos, a statistical method used for visualisation of motifs. The DNA binding motif of the TBP protein is visually represented by a <a class="reference internal" href="gene_expression.html#binding-to-dna"><span class="std std-ref">sequence logo</span></a>. A related visualisation technique was developed by Zhu et al (from my own lab) for identifying <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/27974498">sequence motifs associated with mutation processes</a>.</p>
<p>As the start codon example illustrates, motifs represent a fundamental concept in the description of information encoding by DNA sequences.</p>
</div>
</div>
<div class="section" id="exercises">
<h1><span class="section-number">6.5. </span>Exercises<a class="headerlink" href="#exercises" title="Permalink to this headline">¶</a></h1>
<ol class="arabic">
<li><p>Consider the sequence <code class="docutils literal notranslate"><span class="pre">seq</span></code>. How many different <span class="math notranslate nohighlight">\(k\)</span>-mers are there for <span class="math notranslate nohighlight">\(k=1,2,3\)</span>?</p>
<blockquote>
<div><div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;ACG&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div></blockquote>
</li>
<li><p>For a sequence of length 7, how many <span class="math notranslate nohighlight">\(k\)</span>-mers are there for <span class="math notranslate nohighlight">\(k=1,2,3\)</span>?</p></li>
<li><p>Write an equation for the number of <span class="math notranslate nohighlight">\(k\)</span>-mers in a sequence of length <span class="math notranslate nohighlight">\(n\)</span>. When you set <span class="math notranslate nohighlight">\(n=3, 7\)</span> and <span class="math notranslate nohighlight">\(k=1, 2, 3\)</span> you should get the same answers as above.</p></li>
<li><p>Write an algorithm that produces all the dinucleotides for <code class="docutils literal notranslate"><span class="pre">seq</span></code>.</p></li>
<li><p>Then do it for all the trinucleotides in <code class="docutils literal notranslate"><span class="pre">seq</span></code>.</p></li>
<li><p>The Python standard library has lots of very useful goodies. Investigate the <code class="docutils literal notranslate"><span class="pre">Counter</span></code> class from <code class="docutils literal notranslate"><span class="pre">collections</span></code> and apply it to your result from (5) and (6) to determine the <span class="math notranslate nohighlight">\(k\)</span>-mer counts. (Use google!)</p></li>
</ol>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="gene_expression.html" class="btn btn-neutral float-right" title="6.6. Gene expression in a nutshell" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="encoding_information.html" class="btn btn-neutral float-left" title="6.3. The role of sequence in encoding life" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020-2021, Gavin Huttley.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>