
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>6.9. Finding motifs &#8212; Topics in Bioinformatics</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=fd3f3429" />
    <link rel="stylesheet" type="text/css" href="../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css?v=3ce756a4" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="../_static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="../_static/documentation_options.js?v=06f64241"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/thebelab-helper.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'seqcomp/motif_finding';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="6.10. Measuring information using Shannon’s entropy" href="information.html" />
    <link rel="prev" title="6.8. Experimental procedures for detecting functionally related sequences" href="related_by_function.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Topics in Bioinformatics</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">6. Sequence Comparison</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="outline.html">6.1. Outline</a></li>
<li class="toctree-l2"><a class="reference internal" href="intro.html">6.2. Introduction to Comparison of Biological Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="encoding_information.html">6.3. The role of sequence in encoding life</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_expression.html">6.4. Gene expression in a nutshell</a></li>

<li class="toctree-l2"><a class="reference internal" href="motifs.html">6.6. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a></li>

<li class="toctree-l2"><a class="reference internal" href="related_by_function.html">6.8. Experimental procedures for detecting functionally related sequences</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">6.9. Finding motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html">6.10. Measuring information using Shannon’s entropy</a></li>

<li class="toctree-l2"><a class="reference internal" href="pssm.html">6.12. Position Specific Scoring Matrices (PSSMs)</a></li>

<li class="toctree-l2"><a class="reference internal" href="odds_ratios.html">6.14. Odds-ratio’s</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html">6.15. Dotplot – Alignment of sequences related by descent from a common ancestor</a></li>

<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html">6.17. Global Pairwise Alignment Using Dynamic Programming</a></li>

</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../statistics/index.html">9. Statistical analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/edit/main/seqcomp/motif_finding.rst" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/GavinHuttley/tib/issues/new?title=Issue%20on%20page%20%2Fseqcomp/motif_finding.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/seqcomp/motif_finding.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Finding motifs</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-probability-model-for-generating-sequences">6.9.1. Simple probability model for generating sequences</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pssms-position-specific-scoring-matrices">6.9.2. PSSMs: Position Specific Scoring Matrices</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#formal-definition-of-the-probability-model">6.9.3. Formal definition of the probability model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-worked-example">6.9.3.1. A worked example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pseudo-counts-handling-missing-data">6.9.3.2. Pseudo-counts – handling missing data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pssm-limitations">6.9.3.3. PSSM limitations</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="finding-motifs">
<h1><span class="section-number">6.9. </span>Finding motifs<a class="headerlink" href="#finding-motifs" title="Link to this heading">#</a></h1>
<p>So how do we find occurrences of a motif that is “noisy”? If we use the consensus approach, it’s trivial –  exact string match. For the IUPAC ambiguity case, it’s also quite trivial – we could use a regular expression <a class="footnote-reference brackets" href="#id2" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>, e.g.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">re</span>

<span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s2">&quot;T[CT][ACT][GCT][AC]&quot;</span><span class="p">,</span> <span class="s2">&quot;AGCGTTTCTCAGATGCA&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;TTCTC&#39;]
</pre></div>
</div>
</div>
</div>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id2" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>I’ve used the regular expression module (<code class="docutils literal notranslate"><span class="pre">re</span></code>) that’s distributed with python. Parts of the expression such as <code class="docutils literal notranslate"><span class="pre">[CT]</span></code> means call it a match if we have either a C or a T at this position.</p>
</aside>
</aside>
</aside>
<p>Alternatively, we could use a “probabilistic model” <a class="footnote-reference brackets" href="#id4" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. That’s what we’re going to do now.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">2</a><span class="fn-bracket">]</span></span>
<p>A probabilistic model is one where a specific outcome is quantified via explicit probability calculation.</p>
</aside>
</aside>
</aside>
<section id="simple-probability-model-for-generating-sequences">
<h2><span class="section-number">6.9.1. </span>Simple probability model for generating sequences<a class="headerlink" href="#simple-probability-model-for-generating-sequences" title="Link to this heading">#</a></h2>
<p>Assume the probabilities of nucleotides are <span class="math notranslate nohighlight">\(p_A=0.1\)</span>, <span class="math notranslate nohighlight">\(p_C=0.2\)</span>, <span class="math notranslate nohighlight">\(p_G=0.3\)</span>  and <span class="math notranslate nohighlight">\(p_T=0.4\)</span>. We can calculate the probability of observing the following sequence as</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;AGGTC&quot;</span>
<span class="n">probs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">A</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">C</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">G</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">T</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
<span class="n">p_seq</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>
    <span class="n">p_seq</span> <span class="o">*=</span> <span class="n">probs</span><span class="p">[</span><span class="n">base</span><span class="p">]</span>
<span class="n">p_seq</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>0.00072
</pre></div>
</div>
</div>
</div>
<p>For this probability to be an accurate reflection of the true expected frequency of <code class="docutils literal notranslate"><span class="pre">seq</span></code> requires that nucleotides occur independently of each other. Hence, we can just use the product of the individual nucleotide probabilities. (The alternate is a Markov process with order ≥ 1.)</p>
<p>When we consider modelling a sequence motif, however, we use position specific probabilities, i.e. there would be a distinct <code class="docutils literal notranslate"><span class="pre">probs</span></code> for each position of a motif. This is the foundation of PSSMs.</p>
</section>
<section id="pssms-position-specific-scoring-matrices">
<h2><span class="section-number">6.9.2. </span>PSSMs: Position Specific Scoring Matrices<a class="headerlink" href="#pssms-position-specific-scoring-matrices" title="Link to this heading">#</a></h2>
<p>Among the simplest probabilistic approaches are PSSMs. A PSSM is a matrix of log-odds ratios per position of a sequence motif. They are also referred to as profiles. They provide a means for computing the match odds for any new sequence. They are typically applied to finding transcription factor binding sites (TFBS) but also used to characterise protein domains.</p>
<p>Great, where can I get one? Well … recall JASPAR (the open access database of TFBS profiles) <a class="footnote-reference brackets" href="#id6" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a>.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id6" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">3</a><span class="fn-bracket">]</span></span>
<p><em>Sandelin et al (2004). JASPAR: an open-access database for eukaryotic transcription factor binding profiles. Nucleic Acids Research, 32(90001), 91D–94</em></p>
</aside>
</aside>
</aside>
<p>Below is the TATA box position weight matrix from JASPAR.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">A</span>  <span class="p">[</span> <span class="mi">61</span>  <span class="mi">16</span> <span class="mi">352</span>   <span class="mi">3</span> <span class="mi">354</span> <span class="mi">268</span> <span class="mi">360</span> <span class="mi">222</span> <span class="mi">155</span>  <span class="mi">56</span>  <span class="mi">83</span>  <span class="mi">82</span>  <span class="mi">82</span>  <span class="mi">68</span>  <span class="mi">77</span> <span class="p">]</span>
<span class="n">C</span>  <span class="p">[</span><span class="mi">145</span>  <span class="mi">46</span>   <span class="mi">0</span>  <span class="mi">10</span>   <span class="mi">0</span>   <span class="mi">0</span>   <span class="mi">3</span>   <span class="mi">2</span>  <span class="mi">44</span> <span class="mi">135</span> <span class="mi">147</span> <span class="mi">127</span> <span class="mi">118</span> <span class="mi">107</span> <span class="mi">101</span> <span class="p">]</span>
<span class="n">G</span>  <span class="p">[</span><span class="mi">152</span>  <span class="mi">18</span>   <span class="mi">2</span>   <span class="mi">2</span>   <span class="mi">5</span>   <span class="mi">0</span>  <span class="mi">20</span>  <span class="mi">44</span> <span class="mi">157</span> <span class="mi">150</span> <span class="mi">128</span> <span class="mi">128</span> <span class="mi">128</span> <span class="mi">139</span> <span class="mi">140</span> <span class="p">]</span>
<span class="n">T</span>  <span class="p">[</span> <span class="mi">31</span> <span class="mi">309</span>  <span class="mi">35</span> <span class="mi">374</span>  <span class="mi">30</span> <span class="mi">121</span>   <span class="mi">6</span> <span class="mi">121</span>  <span class="mi">33</span>  <span class="mi">48</span>  <span class="mi">31</span>  <span class="mi">52</span>  <span class="mi">61</span>  <span class="mi">75</span>  <span class="mi">71</span> <span class="p">]</span>
</pre></div>
</div>
<p>In this case, the PSSM has length 15. This motif is on display at <a class="reference external" href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl?ID=MA0108.2&amp;rm=present&amp;collection=CORE">JASPAR</a> (but note their web server can be quite slow).</p>
</section>
<section id="formal-definition-of-the-probability-model">
<h2><span class="section-number">6.9.3. </span>Formal definition of the probability model<a class="headerlink" href="#formal-definition-of-the-probability-model" title="Link to this heading">#</a></h2>
<p>We now layout the equations for computing the probability of a sequence. To start, we make some simplifying assumptions. Namely, we assume a counts matrix reflects the true underlying probabilities of bases per position of a motif; that each base in a motif occurs independently of the other bases. We can compute the probability under the PPM (position probability matrix) model <span class="math notranslate nohighlight">\(M\)</span> of a sequence <span class="math notranslate nohighlight">\(x\)</span> with length <span class="math notranslate nohighlight">\(L\)</span> as:</p>
<div class="math notranslate nohighlight">
\[P(x|M)=\Pi_{i=0}^L p_i(x_i)\]</div>
<p>where <span class="math notranslate nohighlight">\(x_i\)</span> is the base at position <span class="math notranslate nohighlight">\(i\)</span>. This probability can also be referred to as the <em>odds</em> of observing the sequence given the model <span class="math notranslate nohighlight">\(M\)</span>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>the <span class="math notranslate nohighlight">\(\Pi\)</span> symbol means “take the product of the series”. So what <span class="math notranslate nohighlight">\(\sum\)</span> is to addition, <span class="math notranslate nohighlight">\(\Pi\)</span> is for multiplication.</p>
</div>
<section id="a-worked-example">
<h3><span class="section-number">6.9.3.1. </span>A worked example<a class="headerlink" href="#a-worked-example" title="Link to this heading">#</a></h3>
<p>Consider the following hypothetical matrix</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">A</span>  <span class="p">[</span> <span class="mi">61</span>  <span class="mi">16</span> <span class="mi">389</span>   <span class="mi">3</span> <span class="p">]</span>
<span class="n">C</span>  <span class="p">[</span><span class="mi">145</span>  <span class="mi">46</span>   <span class="mi">0</span>  <span class="mi">10</span> <span class="p">]</span>
<span class="n">G</span>  <span class="p">[</span><span class="mi">152</span>  <span class="mi">18</span>   <span class="mi">0</span>   <span class="mi">2</span> <span class="p">]</span>
<span class="n">T</span>  <span class="p">[</span> <span class="mi">31</span> <span class="mi">309</span>   <span class="mi">0</span> <span class="mi">374</span> <span class="p">]</span>
</pre></div>
</div>
<p>which produces the following position specific probability matrix</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">A</span> <span class="p">[</span> <span class="mf">0.16</span>  <span class="mf">0.04</span>  <span class="mf">1.00</span>  <span class="mf">0.01</span> <span class="p">]</span>
<span class="n">C</span> <span class="p">[</span> <span class="mf">0.37</span>  <span class="mf">0.12</span>  <span class="mf">0.00</span>  <span class="mf">0.03</span> <span class="p">]</span>
<span class="n">G</span> <span class="p">[</span> <span class="mf">0.39</span>  <span class="mf">0.05</span>  <span class="mf">0.00</span>  <span class="mf">0.01</span> <span class="p">]</span>
<span class="n">T</span> <span class="p">[</span> <span class="mf">0.08</span>  <span class="mf">0.79</span>  <span class="mf">0.00</span>  <span class="mf">0.96</span> <span class="p">]</span>
</pre></div>
</div>
<p>The conditional probability of the sequence:</p>
<p><em>1.</em> <code class="docutils literal notranslate"><span class="pre">GTAT</span></code>, is <span class="math notranslate nohighlight">\(0.39 \times 0.79 \times 1.0 \times 0.96 \approx 0.3\)</span></p>
<p><em>2.</em> <code class="docutils literal notranslate"><span class="pre">GTCT</span></code>, is <span class="math notranslate nohighlight">\(0.39 \times 0.79 \times 0.0 \times 0.96 = 0.0\)</span></p>
<p>The latter case is potentially real – binding only happens with <code class="docutils literal notranslate"><span class="pre">&quot;A&quot;</span></code> at the 3rd position – but most likely due to a finite sized training data set. In other words, if the training set had been much larger we may have observed the other bases at that position.</p>
</section>
<section id="pseudo-counts-handling-missing-data">
<h3><span class="section-number">6.9.3.2. </span>Pseudo-counts – handling missing data<a class="headerlink" href="#pseudo-counts-handling-missing-data" title="Link to this heading">#</a></h3>
<p>Handling small sample sizes is a substantial problem <a class="footnote-reference brackets" href="#id8" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>. The easiest way to tackle it, which we will use here, is to employ a pseudo-count. A pseudo-count is a “synthetic observation” that is added to all the elements in the counts matrix. It eliminates 0 counts and thus precludes cases such as (2) above, where a sequence is otherwise considered impossible. I’ll illustrate that using the above example.</p>
<aside class="margin sidebar">
<p class="sidebar-title"></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id8" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">4</a><span class="fn-bracket">]</span></span>
<p>When sample sizes are large, the effect of adding a pseudo-count is small.</p>
</aside>
</aside>
</aside>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">numpy</span> <span class="kn">import</span> <span class="n">array</span>

<span class="n">counts</span> <span class="o">=</span> <span class="n">array</span><span class="p">(</span>
    <span class="p">[[</span><span class="mi">61</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">389</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">145</span><span class="p">,</span> <span class="mi">46</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span> <span class="p">[</span><span class="mi">152</span><span class="p">,</span> <span class="mi">18</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">31</span><span class="p">,</span> <span class="mi">309</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">374</span><span class="p">]]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>Then we add our pseudo-count.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">counts</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="n">counts</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>array([[ 62,  17, 390,   4],
       [146,  47,   1,  11],
       [153,  19,   1,   3],
       [ 32, 310,   1, 375]])
</pre></div>
</div>
</div>
</div>
<p>We determine our new columns sums</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">col_sums</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">col_sums</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>array([393, 393, 393, 393])
</pre></div>
</div>
</div>
</div>
<p>and produce a revised position specific probability matrix</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ppm</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">/</span> <span class="n">col_sums</span>
<span class="n">ppm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>array([[0.15776081, 0.043257  , 0.99236641, 0.01017812],
       [0.37150127, 0.11959288, 0.00254453, 0.02798982],
       [0.38931298, 0.04834606, 0.00254453, 0.00763359],
       [0.08142494, 0.78880407, 0.00254453, 0.95419847]])
</pre></div>
</div>
</div>
</div>
<hr class="docutils" />
<p>In order to assess whether a sequence might be consistent with a PSSM, we need a way of scoring them. This is done by converting a the Position specific Probability Matrix (PPM) to a log-odds ratio. In short, we compare the odds of observing a base compared to its odds from a background distribution.</p>
<p>For convenience, define the background distribution of sequence states to be equally frequent. Then the odds ratio is:</p>
<div class="math notranslate nohighlight">
\[OR(x|M)=\Pi_{i=0}^L \frac{p_i(x_i)}{0.25}\]</div>
<p>Which is expressed as a log-odds score <span class="math notranslate nohighlight">\(S\)</span>:</p>
<div class="math notranslate nohighlight">
\[S=\sum_{i=0}^L \log p_i(x_i) - \log 0.25\]</div>
<p>We then interpret the values of <span class="math notranslate nohighlight">\(S\)</span> as</p>
<ul class="simple">
<li><p>if <span class="math notranslate nohighlight">\(S=0\)</span>, the sequence is equally likely in the PSSM and background</p></li>
<li><p>if <span class="math notranslate nohighlight">\(S&lt;0\)</span>, the sequence is less likely under the PSSM than background</p></li>
<li><p>if <span class="math notranslate nohighlight">\(S&gt;0\)</span>, the sequence is more likely under the PSSM than background</p></li>
</ul>
</section>
<section id="pssm-limitations">
<h3><span class="section-number">6.9.3.3. </span>PSSM limitations<a class="headerlink" href="#pssm-limitations" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>if the training data is limited, we need to handle zero counts which may introduce bias</p></li>
<li><p>we assume bases in a sequence are independent of each other</p></li>
</ul>
</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="related_by_function.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">6.8. </span>Experimental procedures for detecting functionally related sequences</p>
      </div>
    </a>
    <a class="right-next"
       href="information.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">6.10. </span>Measuring information using Shannon’s entropy</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-probability-model-for-generating-sequences">6.9.1. Simple probability model for generating sequences</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pssms-position-specific-scoring-matrices">6.9.2. PSSMs: Position Specific Scoring Matrices</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#formal-definition-of-the-probability-model">6.9.3. Formal definition of the probability model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-worked-example">6.9.3.1. A worked example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pseudo-counts-handling-missing-data">6.9.3.2. Pseudo-counts – handling missing data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pssm-limitations">6.9.3.3. PSSM limitations</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gavin Huttley
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2020-2024, Gavin Huttley.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>