

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>6.6. Experimental procedures for detecting functionally related sequences &mdash; Topics in Bioinformatics 2020 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
  <link rel="stylesheet" href="../_static/sphinx-dropdown.css" type="text/css" />
  <link rel="stylesheet" href="../_static/panels-bootstrap.min.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="6.7. Finding motifs" href="motif_finding.html" />
    <link rel="prev" title="6.5. Gene expression in a nutshell" href="gene_expression.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> Topics in Bioinformatics
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">6. Sequence Comparison</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="intro.html">6.1. Introduction to Comparison of Biological Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="encoding_information.html">6.2. The role of sequence in encoding life</a></li>
<li class="toctree-l2"><a class="reference internal" href="motifs.html">6.3. <span class="math notranslate nohighlight">\(k\)</span>-mers and Motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="motifs.html#exercises">6.4. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="gene_expression.html">6.5. Gene expression in a nutshell</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">6.6. Experimental procedures for detecting functionally related sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="motif_finding.html">6.7. Finding motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html">6.8. Measuring information using Shannon’s entropy</a></li>
<li class="toctree-l2"><a class="reference internal" href="information.html#exercises">6.9. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html">6.10. Position Specific Scoring Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="pssm.html#exercises">6.11. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="odds_ratios.html">6.12. Odds-ratio’s</a></li>
<li class="toctree-l2"><a class="reference internal" href="statistics.html">6.13. Refresher on statistical inference</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html">6.14. Dotplot – Alignment of sequences related by descent from a common ancestor</a></li>
<li class="toctree-l2"><a class="reference internal" href="dotplot.html#exercises">6.15. Exercises</a></li>
<li class="toctree-l2"><a class="reference internal" href="pairwise_alignment.html">6.16. Pairwise alignment using dynamic programming</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">7. Index</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="index.html"><span class="section-number">6. </span>Sequence Comparison</a> &raquo;</li>
        
      <li><span class="section-number">6.6. </span>Experimental procedures for detecting functionally related sequences</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/seqcomp/related_by_function.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="experimental-procedures-for-detecting-functionally-related-sequences">
<span id="experimental-data"></span><h1><span class="section-number">6.6. </span>Experimental procedures for detecting functionally related sequences<a class="headerlink" href="#experimental-procedures-for-detecting-functionally-related-sequences" title="Permalink to this headline">¶</a></h1>
<p>Sequences that a possess a shared functional property (e.g. protein binding) will share sequence features in common. For instance, the DNA sequences that are able to bind TBP. What sort of experiments can be conducted that allow identifying whether there is a particular diagnostic DNA sequence motif to which a protein binds? Whatever the nature of the experiment, we need a means for isolating a collection of sequences that are enriched for those that bind to the protein of interest. The experiment also needs to be able to identify individual sequences that have bound. Once that data exists we enter the world of computation and statistical analysis.</p>
<p>There are a variety experimental procedures that can used for this purpose <a class="bibtex reference internal" href="#geertz-2010aa" id="id1">[GM10]</a>. I will discuss just two of those here.</p>
<div class="section" id="selex-systematic-evolution-of-ligands-by-exponential-enrichment">
<span id="index-0"></span><h2><span class="section-number">6.6.1. </span>SELEX – Systematic evolution of ligands by exponential enrichment<a class="headerlink" href="#selex-systematic-evolution-of-ligands-by-exponential-enrichment" title="Permalink to this headline">¶</a></h2>
<p>This procedure is entirely <em>in vitro</em>. The inputs are a substantial amount of enriched protein <a class="footnote-reference brackets" href="#id3" id="id2">1</a> which is bound to a bead. Also required is a library of synthetically produced double stranded DNA of a precise fragment size. These two substrate are then incubated together under conditions favourable to binding of the DNA and protein. By eluting the unbound DNA fragments, you wind up with beads that have bound DNA. Those DNA fragments are then dissociated from the beads and amplified using PCR. This new DNA pool is then reintroduced to bead-bound protein and the process is repeated. (Only a few a rounds are done.) At the end of these iterations, the bound DNA is isolated again and the collection of DNA fragments is sequenced.</p>
<dl class="footnote brackets">
<dt class="label" id="id3"><span class="brackets"><a class="fn-backref" href="#id2">1</a></span></dt>
<dd><p>The requirement for a lot of protein limits the utility of this technique.</p>
</dd>
</dl>
<div class="figure align-default" id="id9">
<a class="reference internal image-reference" href="../_images/selex.png"><img alt="../_images/selex.png" src="../_images/selex.png" style="width: 190.5px; height: 271.0px;" /></a>
<p class="caption"><span class="caption-text">A synthetic sequence assessment procedure.</span><a class="headerlink" href="#id9" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="chip-chromatin-immunoprecipitation">
<span id="index-1"></span><h2><span class="section-number">6.6.2. </span>ChIP – chromatin immunoprecipitation<a class="headerlink" href="#chip-chromatin-immunoprecipitation" title="Permalink to this headline">¶</a></h2>
<p>ChIP-seq is an <em>in vivo</em> process. A precursor for this technique is availability of an antibody with high specificity for the protein of interest. With this in place, the cellular material of interest is exposed to formaldehyde which causes formation of cross linking (via covalent bonds) between DNA and whatever else is bound to it. The cells are then lysed and the DNA protein mix extracted and sheared <a class="footnote-reference brackets" href="#id5" id="id4">2</a> to a size suitable for the sequencing technology that will be used. You then expose the sheared DNA+protein mixture to the antibody and precipitate bound complexes. This step is followed by chemistry that reverses the cross linking, the protein is removed and the collection of DNA fragments are sequenced.</p>
<dl class="footnote brackets">
<dt class="label" id="id5"><span class="brackets"><a class="fn-backref" href="#id4">2</a></span></dt>
<dd><p>Sonication being one approach.</p>
</dd>
</dl>
<div class="figure align-default" id="id10">
<a class="reference internal image-reference" href="../_images/chipseq.png"><img alt="../_images/chipseq.png" src="../_images/chipseq.png" style="width: 283.5px; height: 212.5px;" /></a>
<p class="caption"><span class="caption-text">An empirical survey of naturally occurring DNA <a class="footnote-reference brackets" href="#id7" id="id6">5</a>.</span><a class="headerlink" href="#id10" title="Permalink to this image">¶</a></p>
</div>
<dl class="footnote brackets">
<dt class="label" id="id7"><span class="brackets"><a class="fn-backref" href="#id6">5</a></span></dt>
<dd><p><a class="reference external" href="https://en.wikipedia.org/wiki/ChIP-sequencing">Wikipedia entry</a></p>
</dd>
</dl>
</div>
<div class="section" id="with-the-sequence-data-from-those-experimental-procedures">
<h2><span class="section-number">6.6.3. </span>With the sequence data from those experimental procedures<a class="headerlink" href="#with-the-sequence-data-from-those-experimental-procedures" title="Permalink to this headline">¶</a></h2>
<p>Identifying a dominant motif requires a way of summarising features across a collection of sequences. For instance, we can “align” fragments and employ a majority rule consensus approach. This just picks the most frequent state in a column of <em>aligned</em> sequences.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>01234 &lt;-- the &quot;index&quot; or position
TCAGA
TTCCA
TTCCA
TTTTC
TTTTC

TTCTA &lt;-- the majority consensus
</pre></div>
</div>
<p>Challenges to this approach include handling the case of equally abundant states (a random choice at positions 2 and 3), and masking the possible importance of other states. An alternate approach to handling this issue of multiple characters is to use <a class="reference external" href="https://en.wikipedia.org/wiki/Nucleic_acid_notation">IUPAC ambiguity characters</a> to capture all states at a column.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>01234
TCAGA
TTCCA
TTCCA
TTTTC
TTTTC

TYHBM &lt;-- the IUPAC consensus
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the above, <code class="docutils literal notranslate"><span class="pre">Y</span></code> is either <code class="docutils literal notranslate"><span class="pre">C</span></code> or <code class="docutils literal notranslate"><span class="pre">T</span></code>.</p>
</div>
</div>
<div class="section" id="transformation-of-the-data-for-analysis">
<span id="pssm-origins"></span><h2><span class="section-number">6.6.4. </span>Transformation of the data for analysis<a class="headerlink" href="#transformation-of-the-data-for-analysis" title="Permalink to this headline">¶</a></h2>
<p>From an experimental procedure, we ultimately seek to obtain a curated set of “aligned” sequences.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: center !important; padding: 0 5px;}
.c3table tr > td {text-align: right !important; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7;}
.c3table .index {background: rgba(161, 195, 209, 0.25); font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
</style>
<div class="c3table">
<table>
<caption><span class="cell_title">Hypothetical sequence data from an enrichment experiment</span><br><span class="cell_legend">Position -- alignment position; label -- sequence identifier. The value '.' indicates the nucleotide is identical to that of the first sequence.</span></caption>
<thead class="head_cell">
<th>Label \ Position</th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</thead>
<tbody>
<tr>
<td>seq-0</td>
<td>A</td>
<td>T</td>
<td>T</td>
<td>T</td>
<td>A</td>
<td>T</td>
<td>G</td>
</tr>
<tr>
<td>seq-1</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>A</td>
</tr>
<tr>
<td>seq-2</td>
<td>T</td>
<td>.</td>
<td>A</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>A</td>
</tr>
<tr>
<td>seq-3</td>
<td>T</td>
<td>.</td>
<td>A</td>
<td>A</td>
<td>.</td>
<td>A</td>
<td>.</td>
</tr>
<tr>
<td>seq-4</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>A</td>
<td>.</td>
<td>.</td>
<td>.</td>
</tr>
<tr>
<td>seq-5</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>.</td>
<td>.</td>
<td>.</td>
<td>.</td>
</tr>
<tr>
<td>seq-6</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>.</td>
<td>.</td>
<td>G</td>
<td>.</td>
</tr>
<tr>
<td>seq-7</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>A</td>
<td>.</td>
<td>A</td>
<td>A</td>
</tr>
<tr>
<td>seq-8</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>A</td>
<td>.</td>
<td>.</td>
<td>C</td>
</tr>
<tr>
<td>seq-9</td>
<td>.</td>
<td>.</td>
<td>A</td>
<td>.</td>
<td>T</td>
<td>.</td>
<td>A</td>
</tr>
</tbody>
</table>
<p>
10 rows x 8 columns</p>
</div></div></div>
</div>
<p>This is converted to a table of counts by simply counting occurrences of bases in each alignment column.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: center !important; padding: 0 5px;}
.c3table tr > td {text-align: right !important; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7;}
.c3table .index {background: rgba(161, 195, 209, 0.25); font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
</style>
<div class="c3table">
<table>
<caption><span class="cell_title">PWM</span><br><span class="cell_legend">position specific weights matrix</span></caption>
<thead class="head_cell">
<th>Base \ Position</th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
</thead>
<tbody>
<tr>
<td>T</td>
<td>2</td>
<td>10</td>
<td>1</td>
<td>6</td>
<td>1</td>
<td>5</td>
<td>0</td>
</tr>
<tr>
<td>C</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>A</td>
<td>8</td>
<td>0</td>
<td>9</td>
<td>4</td>
<td>9</td>
<td>4</td>
<td>4</td>
</tr>
<tr>
<td>G</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>5</td>
</tr>
</tbody>
</table>
<p>
4 rows x 8 columns</p>
</div></div></div>
</div>
<hr class="docutils" />
<p class="rubric">Citations</p>
<p id="bibtex-bibliography-seqcomp/related_by_function-0"><dl class="citation">
<dt class="bibtex label" id="geertz-2010aa"><span class="brackets"><a class="fn-backref" href="#id1">GM10</a></span></dt>
<dd><p>Marcel Geertz and Sebastian J Maerkl. Experimental strategies for studying transcription factor-dna binding specificities. <em>Brief Funct Genomics</em>, 9:362–73, 2010. <a class="reference external" href="https://doi.org/10.1093/bfgp/elq023">doi:10.1093/bfgp/elq023</a>.</p>
</dd>
</dl>
</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="motif_finding.html" class="btn btn-neutral float-right" title="6.7. Finding motifs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="gene_expression.html" class="btn btn-neutral float-left" title="6.5. Gene expression in a nutshell" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Gavin Huttley

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(false);
      });
  </script>

  
  
    
   

</body>
</html>