<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>9.2. Case studies &mdash; Topics in Bioinformatics 2023.03.23 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script src="../_static/design-tabs.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="../_static/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="9.3. Distributions and the \(p\)-value" href="distributions.html" />
    <link rel="prev" title="9.1. Evidence based decision making" href="intro.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">9. Statistical analysis</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="intro.html">9.1. Evidence based decision making</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">9.2. Case studies</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#case-study-1-do-nucleotides-occur-randomly-in-a-dna-sequence">9.2.1. Case study 1 – do nucleotides occur randomly in a DNA sequence?</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#step-1-the-biological-problem">9.2.1.1. Step 1 - the biological problem</a></li>
<li class="toctree-l4"><a class="reference internal" href="#step-2-selecting-the-measurement">9.2.1.2. Step 2 - selecting the measurement</a></li>
<li class="toctree-l4"><a class="reference internal" href="#step-3-writing-the-algorithm">9.2.1.3. Step 3 - writing the algorithm</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#case-study-2-evaluate-whether-a-query-sequence-belongs-to-a-pathogen">9.2.2. Case study 2 – evaluate whether a query sequence belongs to a pathogen</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id9">9.2.2.1. Step 1 - the biological problem</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id12">9.2.2.2. Step 2 - selecting the measurement</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id15">9.2.2.3. Step 3 - writing the algorithm</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="distributions.html">9.3. Distributions and the <span class="math notranslate nohighlight">\(p\)</span>-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpretation.html">9.4. Step 4 – Interpretation and looking past the <span class="math notranslate nohighlight">\(p\)</span>-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="wrong_models.html">9.5. Acknowledging the limits of an analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="multiple_testing.html">9.6. Hypothesis testing on genomic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="uses_for_distributions.html">9.7. Beyond hypothesis testing – exploratory data analysis for genomics</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html"><span class="section-number">9. </span>Statistical analysis</a></li>
      <li class="breadcrumb-item active"><span class="section-number">9.2. </span>Case studies</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/statistics/case_studies.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<section id="case-studies">
<span id="id1"></span><h1><span class="section-number">9.2. </span>Case studies<a class="headerlink" href="#case-studies" title="Permalink to this heading"></a></h1>
<p>We will develop a perspective on statistical analysis through addressing two different problems. In these case studies, I’m describing the problem in a very procedural way, focussing on the <a class="reference internal" href="intro.html#essential-steps"><span class="std std-ref">essential steps</span></a> to deliver a <span class="math notranslate nohighlight">\(p\)</span>-value. I am taking this approach because many of the students (and in fact many of the scientists) I have interacted with believe that this is all statistical analysis entails.</p>
<p>The point of this topic is explain <strong>why</strong> this a manifestly superficial approach to statistical analysis. In subsequent sections I develop our interpretation of the analysis results I obtain here.</p>
<section id="case-study-1-do-nucleotides-occur-randomly-in-a-dna-sequence">
<span id="case-study-1"></span><h2><span class="section-number">9.2.1. </span>Case study 1 – do nucleotides occur randomly in a DNA sequence?<a class="headerlink" href="#case-study-1-do-nucleotides-occur-randomly-in-a-dna-sequence" title="Permalink to this heading"></a></h2>
<p>Much of my own research concerns understanding organisation <a class="footnote-reference brackets" href="#id4" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a> of DNA sequences and the mechanisms by which those properties arise. Inevitably, this leads me to asking questions about the role that mutation plays in creating patterns in DNA. This particular case study is focussed on a fundamental property of DNA that has important implications for a large number of fields, including molecular evolutionary analyses such as phylogenetics, and cancer genetics. We now follow the 4-step procedure.</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id4" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">1</a><span class="fn-bracket">]</span></span>
<p>By organisation I mean the ordering of nucleotides in a DNA sequence, e.g. does eukaryotic <a class="reference internal" href="../seqcomp/encoding_information.html#organisation-dna"><span class="std std-ref">DNA organisation</span></a> manifest in periodic base patterns? I consider understanding such properties as essential to understanding how DNA encodes life.</p>
</aside>
</aside>
<section id="step-1-the-biological-problem">
<h3><span class="section-number">9.2.1.1. </span>Step 1 - the biological problem<a class="headerlink" href="#step-1-the-biological-problem" title="Permalink to this heading"></a></h3>
<aside class="sidebar" id="case-study-1-data">
<p class="sidebar-title">Data for case study 1</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">demo</td><td><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="A_dna">A</span></td></tr>
</table>
<p><i>10 DnaSequence</i></p>
</div></div></div>
</div>
<p>A 10bp snippet of the sample sequence. Click to download the <a class="reference download internal" download="" href="../_downloads/3c037ff4c35022eebe9fd7120b97e728/case_study1.fasta"><code class="xref download docutils literal notranslate"><span class="pre">sequence</span> <span class="pre">data</span></code></a>.</p>
</aside>
<p>For the <a class="reference internal" href="#case-study-1-data"><span class="std std-ref">single DNA sequence</span></a>, we want to evaluate whether nucleotides occur randomly along the sequence <a class="footnote-reference brackets" href="#id6" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a>. In other words, given a nucleotide <code class="docutils literal notranslate"><span class="pre">C</span></code>, is its 3’- neighbour a random draw from all 4 nucleotides?</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id6" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">2</a><span class="fn-bracket">]</span></span>
<p>We are referring to nucleotides on a single strand of the double-helix.</p>
</aside>
</aside>
</section>
<section id="step-2-selecting-the-measurement">
<h3><span class="section-number">9.2.1.2. </span>Step 2 - selecting the measurement<a class="headerlink" href="#step-2-selecting-the-measurement" title="Permalink to this heading"></a></h3>
<p>DNA sequences are represented by the <em>discrete series</em> of symbols <a class="footnote-reference brackets" href="#id8" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a> which we can view as categories. As I’m interested in asking whether these symbols occur randomly, my measurement needs to allow me to distinguish between randomness and non-randomness.</p>
<p>In the statistical parlance of categorical analysis, the synonyms for randomness and non-randomness are independence and dependence respectively. These synonyms help me identify the <span class="math notranslate nohighlight">\(\chi^2\)</span> test of independence as the test I will use. (Selecting this statistical test means we have selected a model for the null distribution, but more on that when we get to the Interpretation step.)</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id8" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">3</a><span class="fn-bracket">]</span></span>
<p>These symbols are the letters of the DNA alphabet <span class="incremental">A, C, G, T</span>.</p>
</aside>
</aside>
<p>That all sounds great, but how can I transform my data in such a way that it can be used for this analysis?</p>
</section>
<section id="step-3-writing-the-algorithm">
<h3><span class="section-number">9.2.1.3. </span>Step 3 - writing the algorithm<a class="headerlink" href="#step-3-writing-the-algorithm" title="Permalink to this heading"></a></h3>
<p>In short, we need to produce what’s referred to as a contingency-table. This is just a table of counts of the number of occurrences of each possibility. Consider the nucleotide <code class="docutils literal notranslate"><span class="pre">A</span></code>. For <code class="docutils literal notranslate"><span class="pre">seq</span></code>, how many times does that occur as: <code class="docutils literal notranslate"><span class="pre">AA</span></code>, <code class="docutils literal notranslate"><span class="pre">AC</span></code>, <code class="docutils literal notranslate"><span class="pre">AG</span></code>, <code class="docutils literal notranslate"><span class="pre">AT</span></code>? For the nucleotide <code class="docutils literal notranslate"><span class="pre">C</span></code>, we need to count the number of occurrences of <code class="docutils literal notranslate"><span class="pre">CA</span></code>, <code class="docutils literal notranslate"><span class="pre">CC</span></code>, <code class="docutils literal notranslate"><span class="pre">CG</span></code>, <code class="docutils literal notranslate"><span class="pre">CT</span></code>. Then for <code class="docutils literal notranslate"><span class="pre">G</span></code> and <code class="docutils literal notranslate"><span class="pre">T</span></code>. The pattern here is we need to count dinucleotides!</p>
<p>Awesome, I need to write an algorithm to convert a sequence into dinucleotides. For example, <code class="docutils literal notranslate"><span class="pre">'ATGAAT'</span></code> should be converted to <code class="docutils literal notranslate"><span class="pre">['AT',</span> <span class="pre">'GA',</span> <span class="pre">'AT']</span></code>. The next part is to convert that series of dinucleotides into a <span class="math notranslate nohighlight">\(4\times 4\)</span> table with the row labels corresponding to the first (the 5’-) nucleotide of the dinucleotide and the column labels corresponding to the second (3’-) nucleotide of the dinucleotide.</p>
<p>Applying my algorithm to <code class="docutils literal notranslate"><span class="pre">seq</span></code> I get</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.25); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td class="index"><span class="c3col_left">A</span></td><td><span class="c3col_right">13</span></td><td><span class="c3col_right">19</span></td><td><span class="c3col_right">27</span></td><td><span class="c3col_right">5</span></td></tr>
<tr><td class="index"><span class="c3col_left">C</span></td><td><span class="c3col_right">29</span></td><td><span class="c3col_right">52</span></td><td><span class="c3col_right">19</span></td><td><span class="c3col_right">14</span></td></tr>
<tr><td class="index"><span class="c3col_left">G</span></td><td><span class="c3col_right">11</span></td><td><span class="c3col_right">26</span></td><td><span class="c3col_right">28</span></td><td><span class="c3col_right">13</span></td></tr>
<tr><td class="index"><span class="c3col_left">T</span></td><td><span class="c3col_right">5</span></td><td><span class="c3col_right">15</span></td><td><span class="c3col_right">17</span></td><td><span class="c3col_right">7</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>Further applying the <span class="math notranslate nohighlight">\(\chi^2\)</span> test to this data produces the following <em>result</em>, a <span class="math notranslate nohighlight">\(\chi^2\)</span> <span class="target" id="index-0"></span>test statistic, associated degrees-of-freedom and <span class="math notranslate nohighlight">\(p\)</span>-value. (We will expand on these later.)</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th>chisq</th><th>df</th><th>pvalue</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">22.577</span></td><td><span class="c3col_right">9</span></td><td><span class="c3col_right">0.0072</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>What we do with this result corresponds to the last of the <a class="reference internal" href="intro.html#essential-steps"><span class="std std-ref">essential steps</span></a>, which we do later.</p>
</div>
</section>
</section>
<section id="case-study-2-evaluate-whether-a-query-sequence-belongs-to-a-pathogen">
<span id="case-study-2"></span><h2><span class="section-number">9.2.2. </span>Case study 2 – evaluate whether a query sequence belongs to a pathogen<a class="headerlink" href="#case-study-2-evaluate-whether-a-query-sequence-belongs-to-a-pathogen" title="Permalink to this heading"></a></h2>
<aside class="sidebar" id="data-for-case-study-2">
<p class="sidebar-title">Data for case study 2</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.25); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Reference</span>
</caption>
<thead class="head_cell">
<th>ID</th><th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td class="index"><span class="c3col_left">CP043554.1</span></td><td><span class="c3col_right">399</span></td><td><span class="c3col_right">345</span></td><td><span class="c3col_right">488</span></td><td><span class="c3col_right">321</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.25); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Query samples</span>
</caption>
<thead class="head_cell">
<th>ID</th><th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td class="index"><span class="c3col_left">AP023375.1</span></td><td><span class="c3col_right">394</span></td><td><span class="c3col_right">341</span></td><td><span class="c3col_right">485</span></td><td><span class="c3col_right">320</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP053802.1</span></td><td><span class="c3col_right">398</span></td><td><span class="c3col_right">345</span></td><td><span class="c3col_right">488</span></td><td><span class="c3col_right">322</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP040172.1</span></td><td><span class="c3col_right">397</span></td><td><span class="c3col_right">343</span></td><td><span class="c3col_right">487</span></td><td><span class="c3col_right">322</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP090387.1</span></td><td><span class="c3col_right">396</span></td><td><span class="c3col_right">342</span></td><td><span class="c3col_right">485</span></td><td><span class="c3col_right">320</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP007634.1</span></td><td><span class="c3col_right">397</span></td><td><span class="c3col_right">343</span></td><td><span class="c3col_right">487</span></td><td><span class="c3col_right">322</span></td></tr>
<tr class="ellipsis"><td><span class="c3col_left">...</span></td><td><span class="c3col_right">...</span></td><td><span class="c3col_right">...</span></td><td><span class="c3col_right">...</span></td><td><span class="c3col_right">...</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP016987.1</span></td><td><span class="c3col_right">393</span></td><td><span class="c3col_right">341</span></td><td><span class="c3col_right">484</span></td><td><span class="c3col_right">318</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP025936.1</span></td><td><span class="c3col_right">397</span></td><td><span class="c3col_right">343</span></td><td><span class="c3col_right">488</span></td><td><span class="c3col_right">322</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP010812.1</span></td><td><span class="c3col_right">393</span></td><td><span class="c3col_right">338</span></td><td><span class="c3col_right">483</span></td><td><span class="c3col_right">320</span></td></tr>
<tr><td class="index"><span class="c3col_left">CP003330.1</span></td><td><span class="c3col_right">396</span></td><td><span class="c3col_right">342</span></td><td><span class="c3col_right">486</span></td><td><span class="c3col_right">320</span></td></tr>
<tr><td class="index"><span class="c3col_left">LT992486.1</span></td><td><span class="c3col_right">391</span></td><td><span class="c3col_right">339</span></td><td><span class="c3col_right">486</span></td><td><span class="c3col_right">316</span></td></tr>
</tbody>
</table>
<p>Top 5 and bottom 5 rows from
87 rows x 5 columns</p>
</div></div></div>
</div>
<p>Nucleotide counts from the 16S rRNA gene of the reference and query genomes.</p>
</aside>
<p>The properties of DNA are often used for identification purposes. In this case study we are interested in the detection of the cholera pathogen in amplicon sequence data. A standard approach in analyses of microbial communities is to use the DNA sequence from the <span class="target" id="index-1"></span>16S rRNA gene (hereafter abbreviated 16S) as a species marker. These can be sampled from an environment by PCR amplification and subsequent high-throughput DNA sequencing.</p>
<section id="id9">
<h3><span class="section-number">9.2.2.1. </span>Step 1 - the biological problem<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h3>
<p>Our question is whether a “query” 16S DNA sequence could be a member of <em>Vibrio cholerae</em> (some strains of which cause cholera).</p>
<p>If the query sequence belongs to <em>V. cholerae</em>, then its DNA sequence should be very similar to the reference <a class="footnote-reference brackets" href="#id11" id="id10" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a> <em>V. cholerae</em> pathogen 16S sequence.</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id11" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id10">4</a><span class="fn-bracket">]</span></span>
<p>We use the NCBI defined reference for <em>V. cholerae</em>.</p>
</aside>
</aside>
</section>
<section id="id12">
<h3><span class="section-number">9.2.2.2. </span>Step 2 - selecting the measurement<a class="headerlink" href="#id12" title="Permalink to this heading"></a></h3>
<p>As for <a class="reference internal" href="#case-study-1"><span class="std std-ref">Case study 1</span></a>, because DNA sequences are our basic data type we are dealing with categorical data. We can measure the similarity between DNA sequences in a very large number of ways. One computationally efficient way is to just compare nucleotide counts. If the query belongs to a different species than the reference sequence, we expect the abundance of nucleotides will be less similar than if it is from the same species. In other words, I’m expressing this measurement problem as one of establishing whether a query <em>is not</em> <em>V. cholerae</em>.</p>
<p>Conveniently, this problem can also be evaluated using a <span class="math notranslate nohighlight">\(\chi^2\)</span> test. In this case, it is a homogeneity test <a class="footnote-reference brackets" href="#id14" id="id13" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a>.</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id14" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">5</a><span class="fn-bracket">]</span></span>
<p>Which tests whether the query comes from the same population as the reference.</p>
</aside>
</aside>
</section>
<section id="id15">
<h3><span class="section-number">9.2.2.3. </span>Step 3 - writing the algorithm<a class="headerlink" href="#id15" title="Permalink to this heading"></a></h3>
<p>The first part of the algorithm here is quite simple. For each sequence, we count the number of <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">C</span></code>, <code class="docutils literal notranslate"><span class="pre">G</span></code> and <code class="docutils literal notranslate"><span class="pre">T</span></code>. The result of this is displayed in the case study <a class="reference internal" href="#data-for-case-study-2"><span class="std std-ref">data overview</span></a>. The second part of the algorithm selects the counts for a single query and combines those with the counts from the reference, to produce a <span class="math notranslate nohighlight">\(2 \times 4\)</span> contingency table. For example, with <code class="docutils literal notranslate"><span class="pre">AP023375.1</span></code> as the query we generate</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.25); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th>ID</th><th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td class="index"><span class="c3col_left">CP043554.1</span></td><td><span class="c3col_right">399</span></td><td><span class="c3col_right">345</span></td><td><span class="c3col_right">488</span></td><td><span class="c3col_right">321</span></td></tr>
<tr><td class="index"><span class="c3col_left">AP023375.1</span></td><td><span class="c3col_right">394</span></td><td><span class="c3col_right">341</span></td><td><span class="c3col_right">485</span></td><td><span class="c3col_right">320</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>Performing the <span class="math notranslate nohighlight">\(\chi^2\)</span> homogeneity test produces a <span class="math notranslate nohighlight">\(\chi^2\)</span> <span class="target" id="index-2"></span>test statistic, associated degrees-of-freedom and <span class="math notranslate nohighlight">\(p\)</span>-value.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th>chisq</th><th>df</th><th>pvalue</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.011</span></td><td><span class="c3col_right">3</span></td><td><span class="c3col_right">0.9997</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>This procedure needs to be applied to all the query samples. Again, we will perform the last, interpretation, of out <a class="reference internal" href="intro.html#essential-steps"><span class="std std-ref">essential setsp</span></a> later.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="intro.html" class="btn btn-neutral float-left" title="9.1. Evidence based decision making" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="distributions.html" class="btn btn-neutral float-right" title="9.3. Distributions and the \(p\)-value" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2023, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>