<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>9.4. Step 4 – Interpretation and looking past the \(p\)-value &mdash; Topics in Bioinformatics 2023.03.23 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script src="../_static/design-tabs.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="../_static/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="9.5. Acknowledging the limits of an analysis" href="wrong_models.html" />
    <link rel="prev" title="9.3. Distributions and the \(p\)-value" href="distributions.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">9. Statistical analysis</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="intro.html">9.1. Evidence based decision making</a></li>
<li class="toctree-l2"><a class="reference internal" href="case_studies.html">9.2. Case studies</a></li>
<li class="toctree-l2"><a class="reference internal" href="distributions.html">9.3. Distributions and the <span class="math notranslate nohighlight">\(p\)</span>-value</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">9.4. Step 4 – Interpretation and looking past the <span class="math notranslate nohighlight">\(p\)</span>-value</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#case-study-1">9.4.1. Case Study 1</a></li>
<li class="toctree-l3"><a class="reference internal" href="#case-study-2">9.4.2. Case Study 2</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="wrong_models.html">9.5. Acknowledging the limits of an analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="multiple_testing.html">9.6. Hypothesis testing on genomic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="uses_for_distributions.html">9.7. Beyond hypothesis testing – exploratory data analysis for genomics</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html"><span class="section-number">9. </span>Statistical analysis</a></li>
      <li class="breadcrumb-item active"><span class="section-number">9.4. </span>Step 4 – Interpretation and looking past the <span class="math notranslate nohighlight">\(p\)</span>-value</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/statistics/interpretation.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<section id="step-4-interpretation-and-looking-past-the-pvalue">
<h1><span class="section-number">9.4. </span>Step 4 – Interpretation and looking past the <span class="math notranslate nohighlight">\(p\)</span>-value<a class="headerlink" href="#step-4-interpretation-and-looking-past-the-pvalue" title="Permalink to this heading"></a></h1>
<p>In the <a class="reference internal" href="intro.html#essential-steps"><span class="std std-ref">listing of 4 essential steps</span></a> for data analysis, the final listing concerns interpreting the results of our analysis with respect to our original question.</p>
<section id="case-study-1">
<h2><span class="section-number">9.4.1. </span>Case Study 1<a class="headerlink" href="#case-study-1" title="Permalink to this heading"></a></h2>
<p>I reproduce below the result from analysis of <a class="reference internal" href="case_studies.html#case-study-1"><span class="std std-ref">case study 1</span></a> sequence.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Chisq-test for independence</span>
</caption>
<thead class="head_cell">
<th>chisq</th><th>df</th><th>pvalue</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">22.577</span></td><td><span class="c3col_right">9</span></td><td><span class="c3col_right">0.0072</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>As explained earlier, the <span class="math notranslate nohighlight">\(p\)</span>-value measures the probability that <code class="docutils literal notranslate"><span class="pre">seq</span></code> originated from a process whereby nucleotides occur independently. A <span class="math notranslate nohighlight">\(p\)</span>-value<span class="math notranslate nohighlight">\(\approx\)</span>0.007 indicates this is a rare occurrence. Adopting the conventional threshold of <span class="math notranslate nohighlight">\(\alpha\)</span>=0.05 as our cutoff for declaring we have obtained a “significant” result, we reject that null model as a likely explanation for <code class="docutils literal notranslate"><span class="pre">seq</span></code>. This leads us to the view that nucleotides actually occur in a dependent manner. That seems OK, but we need to challenge our own interpretation.</p>
<p>Simply drawing a conclusion about our data based on the <span class="math notranslate nohighlight">\(p\)</span>-value can be misleading. The amount of data has a direct effect on the power of a statistical test. This translates to <span class="math notranslate nohighlight">\(p\)</span>-values being very sensitive to sample size. For instance, what if we had an enormously long sequence? The <span class="target" id="index-0"></span>statistical power in such a case is so high that we would almost certainly wind up with a tiny <span class="math notranslate nohighlight">\(p\)</span>-value <a class="footnote-reference brackets" href="#id2" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a>. We might be able to convince ourselves that this is not the case here since the sequence is only 600bp long (as distinct from billions of nucleotides long in the case).</p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id2" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>A difference in nucleotide frequency at the 3rd decimal place proved highly significant when the whole vertebrate genome sequences of Mouse and Rat were compared <span id="id3">[<a class="reference internal" href="#id39" title="Gregory M. Cooper, Michael Brudno, Eric A. Stone, Inna Dubchak, Serafim Batzoglou, and Arend Sidow. Characterization of Evolutionary Rates and Constraints in Three Mammalian Genomes. Genome Research, 14(4):539–548, 2004. doi:10.1101/gr.2034704.">CBS+04</a>]</span>! How meaningful is such a result? That’s a good question for which there is no simple answer.</p>
</aside>
</aside>
<p>A more valuable measure is to consider how the data depart from the expectation of the null hypothesis. For categorical data analysis, as we have done here, one very informative quantity are the model <span class="target" id="index-1"></span>residuals. A residual is a scaled measure of the difference between the observed and expected quantities.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.25); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Residuals</span>
</caption>
<thead class="head_cell">
<th></th><th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td class="index"><span class="c3col_left">A</span></td><td><span class="c3col_right">0.1782</span></td><td><span class="c3col_right">-1.0011</span></td><td><span class="c3col_right">1.7219</span></td><td><span class="c3col_right">-1.1510</span></td></tr>
<tr><td class="index"><span class="c3col_left">C</span></td><td><span class="c3col_right">1.4825</span></td><td><span class="c3col_right">1.4470</span></td><td><span class="c3col_right">-2.6494</span></td><td><span class="c3col_right">-0.2130</span></td></tr>
<tr><td class="index"><span class="c3col_left">G</span></td><td><span class="c3col_right">-1.0507</span></td><td><span class="c3col_right">-0.5782</span></td><td><span class="c3col_right">0.8922</span></td><td><span class="c3col_right">0.8981</span></td></tr>
<tr><td class="index"><span class="c3col_left">T</span></td><td><span class="c3col_right">-1.2023</span></td><td><span class="c3col_right">-0.3520</span></td><td><span class="c3col_right">1.0000</span></td><td><span class="c3col_right">0.5352</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>The other merit of examining residuals is that we move from saying that our data differs from the null to demonstrating how that difference occurs. In the above case, the largest absolute residual is -2.7 for the CG dinucleotide. That the residual is negtaive indicates a striking deficit in our data compared to what our null model predicts. This matches up with published data about mutagenesis. Namely, that the CG (or more precisely CpG) dinucleotide is the sequence context within which C becomes methylated (to produce 5-methyl-cytosine, or 5mC). It has been demonstrated that the mutation rate of <span class="math notranslate nohighlight">\(5mC\rightarrow T\)</span> is ~10x the mutation rate of <span class="math notranslate nohighlight">\(C\rightarrow T\)</span>. This mutagenic mechanism thus predicts a deficit of the CpG dinucleotide. So in this case, we have a result which is consistent with evidence from independent studies <span id="id4">[<a class="reference internal" href="#id40" title="M Krawczak, E V Ball, and D N Cooper. Neighboring-nucleotide effects on the rates of germ-line single-base-pair substitution in human genes. Am J Hum Genet, 63(2):474–488, 1998. doi:10.1086/301965.">KBC98</a>]</span>.</p>
</section>
<section id="case-study-2">
<h2><span class="section-number">9.4.2. </span>Case Study 2<a class="headerlink" href="#case-study-2" title="Permalink to this heading"></a></h2>
<p>By now you’ve probably been wondering why I bothered even having the second case study. Fear not, its time has finally come!</p>
<p><a class="reference internal" href="case_studies.html#case-study-2"><span class="std std-ref">Case Study 2</span></a> was selected because it is nearly the reverse of a conventional statistical analysis. Typically, the scientist secretly hopes their statistical analysis will have provided evidence (i.e. a <span class="math notranslate nohighlight">\(p\)</span>-value) that enables them to conclude their null model is unlikely. Thus, their interest in the outcome is maximised when the null is rejected. Recall that our null model for this case study is that a query sample is drawn from the same distribution as our reference sample which is the <em>V. cholerae</em> 16S gene. If the test returns a <span class="math notranslate nohighlight">\(p\)</span>-value&gt;0.05 we face the unpleasant prospect of dealing with an outbreak of cholera. In other words, our interest is maximised when the null is accepted.</p>
<p>There are a multitude of reasons why this approach to detecting cholera is super bad and should never be used. In order to explore these we need to think about what underpins our choices. We could reasonably expect that the DNA sequence of <em>V. cholera</em> encodes the pathogenicity of this bacteria. In order to fit that genomic sequence into a conventional statistical analysis we have adopted an extreme data reduction approach – we selected a single gene and we reduced that gene into just the counts of nucleotides. In doing such a radical data transformation we are assuming that we can uniquely identify a pathogenic cholera bacterium purely from this approach.</p>
<p>The test we employed required the assumption that the genome of cholera exhibits a distinctive nucleotide composition to the sequences of other bacteria. Not only do we require all other bacteria to be markedly different to cholera, we also require all pathogenic cholera to <em>not</em> differ markedly from one another. Ultimately, the mechanism that causes species to differ in the nucleotide abundance is mutagenesis. As our first case study highlighted, mutagenesis is a process in which interactions amongs neighbouring bases can play a profound influence. But the analysis we employed assumed this influence does not exist.</p>
<p>In brief, we should not be confident in the results we obtain from this approach. Clearly, the model used is seriously wrong. In the next section we will demonstrate how model choice can profoundly affect your results.</p>
<hr class="docutils" />
<p class="rubric">Citations</p>
<div class="docutils container" id="id5">
<div class="citation" id="id39" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">CBS+04</a><span class="fn-bracket">]</span></span>
<p>Gregory M. Cooper, Michael Brudno, Eric A. Stone, Inna Dubchak, Serafim Batzoglou, and Arend Sidow. Characterization of Evolutionary Rates and Constraints in Three Mammalian Genomes. <em>Genome Research</em>, 14(4):539–548, 2004. <a class="reference external" href="https://doi.org/10.1101/gr.2034704">doi:10.1101/gr.2034704</a>.</p>
</div>
<div class="citation" id="id40" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id4">KBC98</a><span class="fn-bracket">]</span></span>
<p>M Krawczak, E V Ball, and D N Cooper. Neighboring-nucleotide effects on the rates of germ-line single-base-pair substitution in human genes. <em>Am J Hum Genet</em>, 63(2):474–488, 1998. <a class="reference external" href="https://doi.org/10.1086/301965">doi:10.1086/301965</a>.</p>
</div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="distributions.html" class="btn btn-neutral float-left" title="9.3. Distributions and the \(p\)-value" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="wrong_models.html" class="btn btn-neutral float-right" title="9.5. Acknowledging the limits of an analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2023, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>