<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>9.6. Hypothesis testing on genomic data &mdash; Topics in Bioinformatics 2023.05.01 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/jupyter-sphinx.css" type="text/css" />
      <link rel="stylesheet" href="../_static/thebelab.css" type="text/css" />
      <link rel="stylesheet" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script src="../_static/design-tabs.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../_static/require.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="../_static/require.min.js"></script>
        <script src="../_static/custom.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="9.7. Beyond hypothesis testing – exploratory data analysis for genomics" href="uses_for_distributions.html" />
    <link rel="prev" title="9.5. Acknowledging the limits of an analysis" href="wrong_models.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Topics in Bioinformatics
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">1. Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">2. Prologue</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contribute.html">3. How to Contribute</a></li>
<li class="toctree-l1"><a class="reference internal" href="../notation.html">4. Notation used throughout</a></li>
<li class="toctree-l1"><a class="reference internal" href="../python/index.html">5. Introduction to Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../seqcomp/index.html">6. Sequence Comparison</a></li>
<li class="toctree-l1"><a class="reference internal" href="../molevol/index.html">7. Molecular Evolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cogent3/index.html">8. Molecular Evolutionary Analyses Using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">9. Statistical analysis</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="intro.html">9.1. Evidence based decision making</a></li>
<li class="toctree-l2"><a class="reference internal" href="case_studies.html">9.2. Case studies</a></li>
<li class="toctree-l2"><a class="reference internal" href="distributions.html">9.3. Distributions and the <span class="math notranslate nohighlight">\(p\)</span>-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpretation.html">9.4. Step 4 – Interpretation and looking past the <span class="math notranslate nohighlight">\(p\)</span>-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="wrong_models.html">9.5. Acknowledging the limits of an analysis</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">9.6. Hypothesis testing on genomic data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-bonferroni-correction">9.6.1. The Bonferroni correction</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="uses_for_distributions.html">9.7. Beyond hypothesis testing – exploratory data analysis for genomics</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../version_control/index.html">10. Introduction to version control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../testing/index.html">11. Introduction to Testing in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Topics in Bioinformatics</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html"><span class="section-number">9. </span>Statistical analysis</a></li>
      <li class="breadcrumb-item active"><span class="section-number">9.6. </span>Hypothesis testing on genomic data</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/GavinHuttley/tib/blob/develop/source/statistics/multiple_testing.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="hypothesis-testing-on-genomic-data">
<h1><span class="section-number">9.6. </span>Hypothesis testing on genomic data<a class="headerlink" href="#hypothesis-testing-on-genomic-data" title="Permalink to this heading"></a></h1>
<p>For scientists dealing with genomic data, studies were only a single hypothesis test is performed are rare. Rather, the same hypothesis test may be employed thousands, or hundreds of thousands of times. For instance, Genome Wide Association Studies (or GWAS) evaluate for each Single Nucleotide Polymorphism (or SNP) whether allele counts differ between disease and non-disease samples. With each such test returning a <span class="math notranslate nohighlight">\(p\)</span>-value, the risk of incorrectly rejecting a null hypothesis is greatly amplified.</p>
<p>Let’s define our significance threshold as per the convention, a <span class="math notranslate nohighlight">\(p\)</span>-value<span class="math notranslate nohighlight">\(\le 0.05\)</span>. Drawing on the definition of the <span class="math notranslate nohighlight">\(p\)</span>-value, we expected to obtain a significant result 5% of the time. This means that if we looked at GWAS results from 100 SNPs when there is no genetic basis for a phenotype (i.e. the null hypothesis is true), we expect 5 SNPs to be “significant”. If we looked at 100,000 SNPs, we expect 5000 to be significant.</p>
<p>This is the critical issue confronting genomics studies with very significant practical implications. It not cost-effective to follow up on each such significant result as if it were an equally likely candidate. Instead, we need to employ statistical procedures that adjust for the number of tests that have been performed and increase our confidence that what we select are true positives.</p>
<section id="the-bonferroni-correction">
<h2><span class="section-number">9.6.1. </span>The Bonferroni correction<a class="headerlink" href="#the-bonferroni-correction" title="Permalink to this heading"></a></h2>
<p>This approach to correcting for multiple tests attempts to control the “number” of errors (false discoveries), specifically the probability of seeing 1 or more such errors. This is a strict, very conservative, approach. In applying this, one is prone to accepting the null hypothesis when it’s actually False.</p>
<p>If one conducts <span class="math notranslate nohighlight">\(k\)</span> statistical tests and the desired significance level is <span class="math notranslate nohighlight">\(\alpha\)</span>, then we compute a corrected significance threshold of <span class="math notranslate nohighlight">\(\frac{\alpha}{k}\)</span>. If we set <span class="math notranslate nohighlight">\(\alpha\)</span>=0.05 and <span class="math notranslate nohighlight">\(k\)</span>=100, then the corrected single test threshold is <span class="math notranslate nohighlight">\(\alpha^\prime\)</span>=0.0005.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="wrong_models.html" class="btn btn-neutral float-left" title="9.5. Acknowledging the limits of an analysis" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="uses_for_distributions.html" class="btn btn-neutral float-right" title="9.7. Beyond hypothesis testing – exploratory data analysis for genomics" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2023, Gavin Huttley.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>